{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bc1530f3",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"aggregate_plots\"\n",
    "output: html_document\n",
    "date: \"2024-01-01\"\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1641f9d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(stringr)\n",
    "\n",
    "# Base path for analysis\n",
    "base_path <- \"/workspace/analysis/output/SC3_v3_NextGem_SI_PBMC_10K\"\n",
    "# Target base path\n",
    "target_base_path <- \"/workspace/analysis/downsampled_text_files\"\n",
    "\n",
    "seed_list <- c(\"100\", \"101\", \"102\")\n",
    "\n",
    "# Loop through each variant\n",
    "for (var in c(\"seuratv5.0.2\", \"scanpyv1.9.5\")) {\n",
    "    for (seed in seed_list) {\n",
    "        if (var == \"seuratv5.0.2\") {\n",
    "            input_list <- c(\"input_default\", \"input_seu1\")\n",
    "            short_name <- \"seu\"\n",
    "        } else {\n",
    "            input_list <- c(\"input_default\", \"input_scan1\")\n",
    "            short_name <- \"scan\"\n",
    "        }\n",
    "        \n",
    "        for (input in input_list) {\n",
    "            if (input == \"input_default\") {\n",
    "                cell_list <- c(\"cell_fraction_1_0\", glue::glue(\"cell_fraction_{short_name}1_1_0_vs_{short_name}2_0_64_seed{seed}\"), glue::glue(\"cell_fraction_{short_name}1_1_0_vs_{short_name}2_0_32_seed{seed}\"), glue::glue(\"cell_fraction_{short_name}1_1_0_vs_{short_name}2_0_16_seed{seed}\"), glue::glue(\"cell_fraction_{short_name}1_1_0_vs_{short_name}2_0_08_seed{seed}\"), glue::glue(\"cell_fraction_{short_name}1_1_0_vs_{short_name}2_0_04_seed{seed}\"), glue::glue(\"cell_fraction_{short_name}1_1_0_vs_{short_name}2_0_02_seed{seed}\"), glue::glue(\"cell_fraction_{short_name}1_1_0_vs_{short_name}2_0_01_seed{seed}\"))\n",
    "            } else {\n",
    "                cell_list <- \"cell_fraction_1_0\"\n",
    "            }\n",
    "            \n",
    "            for (cell in cell_list) {\n",
    "                if (cell == \"cell_fraction_1_0\") {\n",
    "                    read_list <- c(glue::glue(\"read_fraction_{short_name}1_1_0_vs_{short_name}2_0_64_seed{seed}\"), glue::glue(\"read_fraction_{short_name}1_1_0_vs_{short_name}2_0_32_seed{seed}\"), glue::glue(\"read_fraction_{short_name}1_1_0_vs_{short_name}2_0_16_seed{seed}\"), glue::glue(\"read_fraction_{short_name}1_1_0_vs_{short_name}2_0_08_seed{seed}\"), glue::glue(\"read_fraction_{short_name}1_1_0_vs_{short_name}2_0_04_seed{seed}\"), glue::glue(\"read_fraction_{short_name}1_1_0_vs_{short_name}2_0_02_seed{seed}\"), glue::glue(\"read_fraction_{short_name}1_1_0_vs_{short_name}2_0_01_seed{seed}\"))\n",
    "                } else {\n",
    "                    read_list <- \"read_fraction_1_0\"\n",
    "                }\n",
    "                \n",
    "                for (read in read_list) {\n",
    "                    path <- file.path(base_path, var, input, \"kb0_28_0\", cell, read, \"stats\")\n",
    "                    \n",
    "                    # Extract information\n",
    "                    seurat_scanpy <- ifelse(str_detect(path, \"seurat\"), \"seurat\", ifelse(str_detect(path, \"scanpy\"), \"scanpy\", NA))\n",
    "                    downsampling_type <- ifelse(str_detect(path, \"/read_fraction.*vs\"), \"read\", ifelse(str_detect(path, \"cell_fraction.*vs\"), \"cell\", NA))\n",
    "                    input_type <- ifelse(str_detect(path, \"input_default\"), \"default\", \"controlled\")\n",
    "                    downsampling_fraction <- str_extract(path, \"(?<=seu2_|scan2_)[0-9]+_[0-9]+\")\n",
    "                    downsampling_seed <- str_extract(path, \"(?<=seed)[0-9]+\")\n",
    "                    \n",
    "                    print(path)\n",
    "                    # Print results\n",
    "                    cat(\"Seurat or Scanpy:\", seurat_scanpy, \"\\n\")\n",
    "                    cat(\"Read or Cell Downsampling:\", downsampling_type, \"\\n\")\n",
    "                    cat(\"Input Default or Controlled:\", input_type, \"\\n\")\n",
    "                    cat(\"Downsampling Fraction:\", downsampling_fraction, \"\\n\")\n",
    "                    cat(\"Downsampling Seed:\", downsampling_seed, \"\\n\")\n",
    "                    \n",
    "                    # Construct the target directory path\n",
    "                    target_path <- file.path(target_base_path, downsampling_type, seurat_scanpy, glue::glue(\"input_{input_type}\"), glue::glue(\"fraction_{downsampling_fraction}\"), glue::glue(\"seed_{downsampling_seed}\"))\n",
    "                \n",
    "                    # Ensure the target directory exists\n",
    "                    if (!dir.exists(target_path)) {\n",
    "                        dir.create(target_path, recursive = TRUE)\n",
    "                    }\n",
    "                    \n",
    "                    files <- list.files(path = source_dir, full.names = TRUE)\n",
    "\n",
    "                    sapply(files, function(file) {file.copy(file, target_dir)})\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e1da282",
   "metadata": {},
   "outputs": [],
   "source": [
    "if (!requireNamespace(\"reticulate\", quietly = TRUE)) remotes::install_version(\"reticulate\", version = \"1.34.0\", upgrade = \"never\")\n",
    "\n",
    "using_colab <- reticulate::py_run_string(\"\n",
    "try:\n",
    "    import google.colab\n",
    "    using_colab = True\n",
    "except ImportError:\n",
    "    using_colab = False\n",
    "using_colab\n",
    "\")$using_colab\n",
    "\n",
    "if (using_colab) {\n",
    "    system(\"git clone https://github.com/josephrich98/scrnaseq_packages_and_versioning.git\", intern = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2541bf83",
   "metadata": {},
   "source": [
    "Select yaml file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "848bad41",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml_file <- \"Fig2_Supp_Fig11_Supp_Fig12\"  # Fig2_Supp_Fig11_Supp_Fig12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f95fe19a",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_version_for_download <- gsub(\"_\", \".\", seurat_version)\n",
    "scanpy_version_for_download <- gsub(\"_\", \".\", scanpy_version)\n",
    "\n",
    "if (using_colab) {\n",
    "    py_command <- sprintf(\"import subprocess; subprocess.run(['pip', 'install', 'scanpy==%s', 'python-igraph==0.10.8', 'leidenalg==0.10.1', 'anndata==0.10.2', 'hdf5plugin==4.2.0', 'kb-python==0.27.3', 'umap-learn==0.5.2', 'louvain==0.8.1', 'git+https://github.com/has2k1/scikit-misc.git@269f61e'])\", scanpy_version_for_download)\n",
    "    \n",
    "    reticulate::py_run_string(py_command)\n",
    "}\n",
    "\n",
    "if (!requireNamespace(\"remotes\", quietly = TRUE)) install.packages(\"remotes\")\n",
    "\n",
    "if (!requireNamespace(\"tidyverse\", quietly = TRUE)) remotes::install_version(\"tidyverse\", version = \"2.0.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"rmarkdown\", quietly = TRUE)) remotes::install_version(\"rmarkdown\", version = \"2.25\", upgrade = \"never\")\n",
    "\n",
    "if (!requireNamespace(\"igraph\", quietly = TRUE)) pak::pak(\"igraph/rigraph\")\n",
    "if (!requireNamespace(\"Seurat\", quietly = TRUE)) remotes::install_version(\"Seurat\", version = seurat_version_for_download, upgrade = \"never\")\n",
    "if (!requireNamespace(\"Matrix\", quietly = TRUE)) remotes::install_version(\"Matrix\", version = \"1.6.4\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"patchwork\", quietly = TRUE)) remotes::install_version(\"patchwork\", version = \"1.1.3\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"eulerr\", quietly = TRUE)) remotes::install_version(\"eulerr\", version = \"7.0.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"scattermore\", quietly = TRUE)) remotes::install_version(\"scattermore\", version = \"1.2\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"assertthat\", quietly = TRUE)) remotes::install_version(\"assertthat\", version = \"0.2.1\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"pheatmap\", quietly = TRUE)) remotes::install_version(\"pheatmap\", version = \"1.0.12\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggforce\", quietly = TRUE)) remotes::install_version(\"ggforce\", version = \"0.4.1\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggplotify\", quietly = TRUE)) remotes::install_version(\"ggplotify\", version = \"0.1.2\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"mclust\", quietly = TRUE)) remotes::install_version(\"mclust\", version = \"6.0.1\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggalluvial\", quietly = TRUE)) remotes::install_version(\"ggalluvial\", version = \"0.12.5\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"UpSetR\", quietly = TRUE)) remotes::install_version(\"UpSetR\", version = \"1.4.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggpointdensity\", quietly = TRUE)) remotes::install_version(\"ggpointdensity\", version = \"0.1.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"dbscan\", quietly = TRUE)) remotes::install_version(\"dbscan\", version = \"1.1.12\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"presto\", quietly = TRUE)) remotes::install_github(\"immunogenomics/presto@31dc97f\", upgrade = \"never\")\n",
    "\n",
    "\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE)) remotes::install_version(\"BiocManager\", version = \"1.30.22\", upgrade = \"never\")\n",
    "bioconductor_version <- \"3.18\"\n",
    "\n",
    "if (!requireNamespace(\"BUSpaRse\", quietly = TRUE)) BiocManager::install(\"BUSpaRse\", version = bioconductor_version, update = FALSE)\n",
    "if (!requireNamespace(\"DropletUtils\", quietly = TRUE)) BiocManager::install(\"DropletUtils\", version = bioconductor_version, update = FALSE)\n",
    "if (!requireNamespace(\"biomaRt\", quietly = TRUE)) BiocManager::install(\"biomaRt\", version = bioconductor_version, update = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aff35e3",
   "metadata": {},
   "source": [
    "Load contents of yaml file into global R environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec1c5670",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml_dir <- glue::glue(\"{dirname(getwd())}/yaml\")\n",
    "\n",
    "yaml_file_path <- glue::glue(\"{yaml_dir}/{yaml_file}.yaml\")\n",
    "\n",
    "config <- yaml::read_yaml(yaml_file_path)\n",
    "\n",
    "for (name in names(config)) {\n",
    "    assign(name, config[[name]], envir = .GlobalEnv)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97d3c7a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "Sys.setenv(RETICULATE_PYTHON = paste(\"/home/rstudio/.conda/envs\", conda_env, \"bin/python3.9\", sep = \"/\"))\n",
    "library(reticulate)\n",
    "use_condaenv(conda_env)\n",
    "library(tidyverse)\n",
    "library(glue)\n",
    "theme_set(theme_bw())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "121f5902",
   "metadata": {},
   "outputs": [],
   "source": [
    "py_run_string('import sys\n",
    "sys.path.append(f\"{r.project_base_path}/scripts\")\n",
    "from download_data import *\n",
    "\n",
    "if r.download_data:\n",
    "    download_and_extract(r.doi, r.data_name_from_download, r.data_path_root)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37882a79",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_stats_filepath_from_download <- function(set, frac_str, package, type_downsampled, seed) {\n",
    "    return(glue('{data_path_root}/{type_downsampled}/{package}/input_{input_type}/fraction_{frac_str}/seed_{seed}/{set$file_name}'))\n",
    "}\n",
    "\n",
    "get_stats_filepath <- function(set, frac_str, package, type_downsampled, seed) {\n",
    "    if (package == \"seu\") {\n",
    "        package_full_name <- \"Seurat\"\n",
    "        package_version <- seurat_version\n",
    "    } else if (package == \"scan\") {\n",
    "        package_full_name <- \"Scanpy\"\n",
    "        package_version <- scanpy_version\n",
    "    }\n",
    "\n",
    "    if (type_downsampled == \"read\") {\n",
    "        downsample_string <- glue(\"cell_fraction_1_0/read_fraction_{package}1_1_0_vs_{package}2_{frac_str}_seed{seed}\")\n",
    "    } else if (type_downsampled == \"cell\") {\n",
    "        downsample_string <- glue(\"cell_fraction_{package}1_1_0_vs_{package}2_{frac_str}/read_fraction_1_0_seed{seed}\")\n",
    "    }\n",
    "\n",
    "    if (set$file_name == \"de_stats.txt\" && set$statistic_name != \"Marker Genes Jaccard\") {\n",
    "        data_input_for_filepath <- glue(\"{package}1\")\n",
    "    } else {\n",
    "        data_input_for_filepath <- \"default\"\n",
    "    }\n",
    "\n",
    "    return(glue(\"{project_base_path}/output/{data_name}/{package_full_name}v{package_version}/input_{data_input_for_filepath}/{matrix_generation}/{downsample_string}/stats/{set$file_name}\"))\n",
    "}\n",
    "\n",
    "find_intersection <- function(df, y_value) {\n",
    "    interpolated <- approx(df$frac, df$value, xout = seq(min(df$frac), max(df$frac), length.out = 1000))\n",
    "\n",
    "    intersect_indices <- which(abs(interpolated$y - y_value) < 0.01)\n",
    "\n",
    "    if (length(intersect_indices) == 0) {\n",
    "        return(min(frac_list))\n",
    "    } else if ((set$ideal_value == 1 && interpolated$y[1] > interpolated$y[which.min(abs(interpolated$y - y_value))]) || (set$ideal_value == 0 && interpolated$y[1] < interpolated$y[which.min(abs(interpolated$y - y_value))])) {\n",
    "        return(min(frac_list))\n",
    "    }\n",
    "    else{\n",
    "        x_intersect <- interpolated$x[which.min(abs(interpolated$y - y_value))]\n",
    "        return(x_intersect)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d35632f8",
   "metadata": {},
   "source": [
    "List of all metrics to extract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e021f592",
   "metadata": {},
   "outputs": [],
   "source": [
    "variable_sets <- list(\n",
    "    list(\n",
    "        statistic_name = \"Cells jaccard\",\n",
    "        file_name = \"euler_stats_afterQC.txt\",\n",
    "        extraction_phrase = \"Cells Jaccard: \",\n",
    "        baseline_value_seurat_vs_scanpy_default = 1,\n",
    "        baseline_value_seurat_vs_scanpy_same_input = NA,\n",
    "        ideal_value = 1,\n",
    "        seurat_noise_value = 0,\n",
    "        scanpy_noise_value = 0,\n",
    "        output_plot_name = \"cells.tiff\"\n",
    "    ),\n",
    "    list(\n",
    "        statistic_name = \"Genes jaccard\",\n",
    "        file_name = \"euler_stats_afterQC.txt\",\n",
    "        extraction_phrase = \"Genes Jaccard: \",\n",
    "        baseline_value_seurat_vs_scanpy_default = 1,\n",
    "        baseline_value_seurat_vs_scanpy_same_input = NA,\n",
    "        ideal_value = 1,\n",
    "        seurat_noise_value = 0,\n",
    "        scanpy_noise_value = 0,\n",
    "        output_plot_name = \"genes.tiff\"\n",
    "    ),\n",
    "    list(\n",
    "        statistic_name = \"HVGs jaccard\",\n",
    "        file_name = \"euler_stats_afterQC.txt\",\n",
    "        extraction_phrase = \"HVGs Jaccard: \",\n",
    "        baseline_value_seurat_vs_scanpy_default = 0.22249151720795,\n",
    "        baseline_value_seurat_vs_scanpy_same_input = NA,\n",
    "        ideal_value = 1,\n",
    "        seurat_noise_value = 0,\n",
    "        scanpy_noise_value = 0,\n",
    "        output_plot_name = \"hvgs.tiff\"\n",
    "    ),\n",
    "    list(\n",
    "        statistic_name = \"Mean ΔPC1-3 loadings\",\n",
    "        file_name = \"pca_knn_clustering_umap_stats.txt\",\n",
    "        extraction_phrase = \"Mean loading difference of PC1-3: \",\n",
    "        baseline_value_seurat_vs_scanpy_default = 0.417599195296815,\n",
    "        baseline_value_seurat_vs_scanpy_same_input = NA,  # 0.476963845838222,\n",
    "        ideal_value = 0,\n",
    "        seurat_noise_value = 0,\n",
    "        scanpy_noise_value = 0,\n",
    "        output_plot_name = \"pca.tiff\"\n",
    "    ),\n",
    "    list(\n",
    "        statistic_name = \"Median log SNN degree ratio\",\n",
    "        file_name = \"pca_knn_clustering_umap_stats.txt\",\n",
    "        extraction_phrase = \"Median magnitude of log degree ratio of SNN: \",\n",
    "        baseline_value_seurat_vs_scanpy_default = 2.05889368905357,\n",
    "        baseline_value_seurat_vs_scanpy_same_input = NA,  # 2.04306872189189,\n",
    "        ideal_value = 0,\n",
    "        seurat_noise_value = 0,\n",
    "        scanpy_noise_value = 0,\n",
    "        output_plot_name = \"knn.tiff\"\n",
    "    ),\n",
    "    list(\n",
    "        statistic_name = \"Clustering ARI\",\n",
    "        file_name = \"pca_knn_clustering_umap_stats.txt\",\n",
    "        extraction_phrase = \"Adjusted Rand index between clusters: \",\n",
    "        baseline_value_seurat_vs_scanpy_default = 0.706349552155871,\n",
    "        baseline_value_seurat_vs_scanpy_same_input = NA,  # 0.87916666950325,\n",
    "        ideal_value = 1,\n",
    "        seurat_noise_value = 0,\n",
    "        scanpy_noise_value = 0.8870013,  \n",
    "        output_plot_name = \"clustering.tiff\"\n",
    "    ),\n",
    "    list(\n",
    "        statistic_name = \"Median jaccard of UMAP KNN\",\n",
    "        file_name = \"pca_knn_clustering_umap_stats.txt\",\n",
    "        extraction_phrase = \"Median jaccard of UMAP KNN: \",\n",
    "        baseline_value_seurat_vs_scanpy_default = 0.0638297872340425,\n",
    "        baseline_value_seurat_vs_scanpy_same_input = NA,  # 0.204819277108434,\n",
    "        ideal_value = 1,\n",
    "        seurat_noise_value = 0.4084507042253520,\n",
    "        scanpy_noise_value = 0.449275362318841,\n",
    "        output_plot_name = \"umap.tiff\"\n",
    "    ),\n",
    "    list(\n",
    "        statistic_name = \"Marker Genes Jaccard\",\n",
    "        file_name = \"de_stats.txt\",\n",
    "        extraction_phrase = \"Marker Genes Jaccard: \",\n",
    "        baseline_value_seurat_vs_scanpy_default = 0.475223613595707,\n",
    "        baseline_value_seurat_vs_scanpy_same_input = NA,  # 0.475609756097561,\n",
    "        ideal_value = 1,\n",
    "        seurat_noise_value = 0,\n",
    "        scanpy_noise_value = 0,\n",
    "        output_plot_name = \"marker_genes.tiff\"\n",
    "    ),\n",
    "    list(\n",
    "        statistic_name = \"Markers Jaccard\",\n",
    "        file_name = \"de_stats.txt\",\n",
    "        extraction_phrase = \"Markers Jaccard: \",\n",
    "        baseline_value_seurat_vs_scanpy_default = NA,\n",
    "        baseline_value_seurat_vs_scanpy_same_input = 0.0498570681569628,\n",
    "        ideal_value = 1,\n",
    "        seurat_noise_value = 0,\n",
    "        scanpy_noise_value = 0,\n",
    "        output_plot_name = \"markers.tiff\"\n",
    "    ),\n",
    "    list(\n",
    "        statistic_name = \"logFC CCC\",\n",
    "        file_name = \"de_stats.txt\",\n",
    "        extraction_phrase = \"logFC CCC: \",\n",
    "        baseline_value_seurat_vs_scanpy_default = NA,\n",
    "        baseline_value_seurat_vs_scanpy_same_input = 0.443526040509361,\n",
    "        ideal_value = 1,\n",
    "        seurat_noise_value = 0,\n",
    "        scanpy_noise_value = 0,\n",
    "        output_plot_name = \"logfc.tiff\"\n",
    "    ),\n",
    "    list(\n",
    "        statistic_name = \"Adj p-value flip rate\",\n",
    "        file_name = \"de_stats.txt\",\n",
    "        extraction_phrase = \"Adjusted p value, fraction that flipped across 0.05 threshold: \",\n",
    "        baseline_value_seurat_vs_scanpy_default = NA,\n",
    "        baseline_value_seurat_vs_scanpy_same_input = 0.130951199338296,\n",
    "        ideal_value = 0,\n",
    "        seurat_noise_value = 0,\n",
    "        scanpy_noise_value = 0,\n",
    "        output_plot_name = \"pvaladj.tiff\"\n",
    "    )\n",
    ") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "794403c3",
   "metadata": {},
   "source": [
    "Downsample line plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e59c1cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "if (type_downsampled == \"cell\") {\n",
    "    variable_sets <- variable_sets[2:8]\n",
    "}\n",
    "\n",
    "statistic_names_total <- unlist(lapply(variable_sets, function(x) x$statistic_name))\n",
    "\n",
    "# variable_sets <- list(\n",
    "#     list(\n",
    "#         statistic_name = \"Mean ΔPC1-3 loadings\",\n",
    "#         file_name = \"pca_knn_clustering_umap_stats.txt\",\n",
    "#         extraction_phrase = \"Mean loading difference of PC1-3: \",\n",
    "#         baseline_value_seurat_vs_scanpy_default = 0.411333932197237,\n",
    "#         baseline_value_seurat_vs_scanpy_same_input = 0.704549354155186,\n",
    "#         ideal_value = 0,\n",
    "#         seurat_noise_value = 0,\n",
    "#         scanpy_noise_value = 0,\n",
    "#         output_plot_name = \"pca.tiff\"\n",
    "#     )\n",
    "# )\n",
    "\n",
    "if (type_downsampled == \"cell\") {\n",
    "    output_path_base <- glue(\"{project_base_path}/output/{data_name}/aggregate_plots/across_downsampled_cells\")\n",
    "} else if (type_downsampled == \"read\") {\n",
    "    output_path_base <- glue(\"{project_base_path}/output/{data_name}/aggregate_plots/across_downsampled_reads\")\n",
    "}\n",
    "\n",
    "if (margin_correct == 0) {\n",
    "    output_stat_filepath <- glue(\"{output_path_base}/intersection_stats.txt\")\n",
    "} else {\n",
    "    output_stat_filepath <- glue(\"{output_path_base}/intersection_stats_margin{margin_correct}.txt\")\n",
    "}\n",
    "\n",
    "if (file.exists(output_stat_filepath)) {\n",
    "    sink(file = output_stat_filepath, append = FALSE)\n",
    "    sink()\n",
    "}\n",
    "\n",
    "\n",
    "for (set in variable_sets) {\n",
    "    results_df <- data.frame(frac = numeric(), value_group1 = numeric(), value_group2 = numeric(), seed = numeric())\n",
    "\n",
    "    output_plot_filepath <- glue(\"{output_path_base}/{set$output_plot_name}\")\n",
    "\n",
    "    for (package in c(\"seu\", \"scan\")) {\n",
    "        for (frac in frac_list) {\n",
    "            for (seed in downsampled_seeds_list) {\n",
    "                frac_str <- gsub(\"\\\\.\", \"_\", as.character(frac)) # fraction of reads after downsampling, as string\n",
    "    \n",
    "                if (download_data) {\n",
    "                    stats_path <- get_stats_filepath_from_download(set, frac_str = frac_str, package = package, type_downsampled = type_downsampled, seed = seed)\n",
    "                } else {\n",
    "                    stats_path <- get_stats_filepath(set, frac_str = frac_str, package = package, type_downsampled = type_downsampled, seed = seed)\n",
    "                }\n",
    "    \n",
    "                # Read the file lines\n",
    "                lines <- readLines(stats_path)\n",
    "    \n",
    "                value_line <- grep(set$extraction_phrase, lines, value = TRUE)\n",
    "    \n",
    "                # Extract the numerical variance using string manipulation\n",
    "                value <- as.numeric(str_extract(value_line, \"(?<=: )\\\\d+(\\\\.\\\\d+)?(e[-+]?\\\\d+)?\"))\n",
    "    \n",
    "                if (package == \"seu\") {\n",
    "                    new_row <- data.frame(frac = frac, value_group1 = value, value_group2 = NA, seed = seed)\n",
    "                } else if (package == \"scan\") {\n",
    "                    new_row <- data.frame(frac = frac, value_group1 = NA, value_group2 = value, seed = seed)\n",
    "                }\n",
    "                # Add the results to the data frame\n",
    "                results_df <- rbind(results_df, new_row)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "\n",
    "    results_df2 <- results_df %>%\n",
    "        distinct() %>%\n",
    "        group_by(frac) %>%\n",
    "        summarize(\n",
    "            value_group1_mean = mean(value_group1, na.rm = TRUE),\n",
    "            value_group2_mean = mean(value_group2, na.rm = TRUE),\n",
    "            sd_group1 = sd(value_group1, na.rm = TRUE),\n",
    "            sd_group2 = sd(value_group2, na.rm = TRUE)\n",
    "        ) %>%\n",
    "        dplyr::rename(\n",
    "            value_group1 = value_group1_mean,\n",
    "            value_group2 = value_group2_mean\n",
    "        )\n",
    "    \n",
    "    long_df <- results_df2 %>%\n",
    "        pivot_longer(\n",
    "            cols = c(value_group1, value_group2),\n",
    "            names_to = \"variable\",\n",
    "            values_to = \"value\"\n",
    "        ) %>%\n",
    "        mutate(\n",
    "            sd = case_when(\n",
    "              variable == \"value_group1\" ~ sd_group1,\n",
    "              variable == \"value_group2\" ~ sd_group2\n",
    "            )\n",
    "        ) %>%\n",
    "        select(-sd_group1, -sd_group2) \n",
    "    \n",
    "    baseline_rows <- tibble(\n",
    "        frac = c(1.0, 1.0),\n",
    "        variable = c(\"value_group1\", \"value_group2\"),\n",
    "        value = c(ifelse(set$seurat_noise_value == 0, set$ideal_value, set$seurat_noise_value), ifelse(set$scanpy_noise_value == 0, set$ideal_value, set$scanpy_noise_value))\n",
    "    )\n",
    "\n",
    "    long_df <- bind_rows(long_df, baseline_rows) %>% mutate(sd = ifelse(is.na(sd), 0, sd))\n",
    "\n",
    "    y_axis_max <- max(1, set$baseline_value_seurat_vs_scanpy_default, set$baseline_value_seurat_vs_scanpy_same_input, max(long_df$value, na.rm = TRUE), na.rm = TRUE)\n",
    "    \n",
    "    if (set$statistic_name == \"logFC CCC\") {\n",
    "        y_axis_min <- -1\n",
    "    } else {\n",
    "        y_axis_min <- 0\n",
    "    }\n",
    "    \n",
    "    long_df <- long_df %>%\n",
    "        mutate(ymin = pmax(value - sd, y_axis_min),\n",
    "               ymax = pmin(value + sd, y_axis_max))\n",
    "\n",
    "    p <- ggplot(long_df, aes(x = frac, y = value, group = variable, color = variable)) +\n",
    "        geom_line(linewidth = 1.3) +\n",
    "        geom_ribbon(aes(ymin = ymin, ymax = ymax, fill = variable), alpha = 0.3, color = NA) +\n",
    "        scale_color_manual(values = c(\"value_group1\" = \"#D55E00\", \"value_group2\" = \"#56B4E9\"), labels = c(\"Seurat\", \"Scanpy\")) +\n",
    "        scale_fill_manual(values = c(\"value_group1\" = \"#D55E00\", \"value_group2\" = \"#56B4E9\"), labels = c(\"Seurat\", \"Scanpy\"), guide = FALSE) +\n",
    "        labs(color = \"variable\") +\n",
    "        theme_minimal() +\n",
    "        theme(\n",
    "            text = element_text(family = \"Arial\"),\n",
    "            legend.position = \"none\",\n",
    "            axis.text.x = element_text(angle = 45, hjust = 1, size = rel(0.55)),\n",
    "            axis.text.y = element_text(size = rel(1)), # Increase axis tick labels size\n",
    "            axis.title = element_text(size = rel(1.45))\n",
    "        ) +\n",
    "        scale_y_continuous(\n",
    "            limits = c(y_axis_min, y_axis_max), # Setting y-axis limits from 0 to 1\n",
    "            breaks = seq(y_axis_min, y_axis_max, by = 0.2), # Major ticks every 0.2\n",
    "            minor_breaks = seq((y_axis_min + 0.1), (y_axis_max - 0.1), by = 0.2)\n",
    "        ) + # Minor ticks at 0.1, 0.3, 0.5, 0.7, 0.9\n",
    "        scale_x_continuous(\n",
    "            breaks = c(frac_list[2:length(frac_list)], 1.0) # Set the breaks at specified points\n",
    "        ) + \n",
    "        coord_cartesian(xlim = c(0, 1), ylim = c(0, y_axis_max)) +\n",
    "        xlab(glue(\"Fraction Downsampled ({type_downsampled}s)\")) +\n",
    "        ylab(set$statistic_name)\n",
    "    \n",
    "    if (set$statistic_name == \"Median log SNN degree ratio\" && type_downsampled == \"cell\") {\n",
    "        p <- p +\n",
    "            scale_y_continuous(\n",
    "                limits = c(y_axis_min, y_axis_max), # Setting y-axis limits from 0 to 1\n",
    "                breaks = seq(y_axis_min, y_axis_max, by = 1), # Major ticks every 0.2\n",
    "                minor_breaks = seq((y_axis_min + 0.5), (y_axis_max - 0.5), by = 0.5)\n",
    "            )\n",
    "    }\n",
    "\n",
    "    if (!is.na(set$baseline_value_seurat_vs_scanpy_default)) {\n",
    "        p <- p +\n",
    "            geom_hline(aes(yintercept = set$baseline_value_seurat_vs_scanpy_default), color = \"black\", linetype = \"dashed\", linewidth = 0.5) # +\n",
    "        # annotate(\"text\", x = 1, y = (set$baseline_value_seurat_vs_scanpy_default + 0.1), label = glue(\"Seurat vs Scanpy, full-size datasets\"), hjust = 1, color = \"black\") +\n",
    "    }\n",
    "\n",
    "    if (!is.na(set$baseline_value_seurat_vs_scanpy_same_input)) {\n",
    "        p <- p +\n",
    "            geom_hline(aes(yintercept = set$baseline_value_seurat_vs_scanpy_same_input), color = \"black\", linetype = \"dashed\", linewidth = 0.5) # +\n",
    "        # annotate(\"text\", x = 1, y = (set$baseline_value_seurat_vs_scanpy_same_input + 0.1), label = glue(\"Seurat vs Scanpy, full-size datasets\"), hjust = 1, color = \"black\") +\n",
    "    }\n",
    "\n",
    "    # if (set$seurat_noise_value != 0) {\n",
    "    #     if ((set$scanpy_noise_value / set$seurat_noise_value) > 0.9 && (set$scanpy_noise_value / set$seurat_noise_value) < 1.1) {\n",
    "    #         noise_color <- \"gray30\"\n",
    "    #         label <- \"Inherent noise\"\n",
    "    #     } else {\n",
    "    #         noise_color <- \"#D55E00\"\n",
    "    #         label <- \"Inherent noise (Seurat)\"\n",
    "    #     }\n",
    "    #     p <- p +\n",
    "    #         # annotate(\"text\", x = 1, y = (set$noise_value + 0.1), label = label, hjust = 1, color = noise_color) +\n",
    "    #         geom_hline(aes(yintercept = set$seurat_noise_value), color = noise_color, linetype = \"solid\", size = 0.5)\n",
    "    # }\n",
    "    # \n",
    "    # if (set$scanpy_noise_value != 0) {\n",
    "    #     if (!(set$seurat_noise_value / set$scanpy_noise_value) > 0.9 && (set$seurat_noise_value / set$scanpy_noise_value) < 1.1) {\n",
    "    #         p <- p +\n",
    "    #             # annotate(\"text\", x = 1, y = (set$noise_value + 0.1), label = \"Inherent noise (Scanpy)\", hjust = 1, color = \"#56B4E9\") +\n",
    "    #             geom_hline(aes(yintercept = set$scanpy_noise_value), color = \"#56B4E9\", linetype = \"solid\", size = 0.5)\n",
    "    #     }\n",
    "    # }\n",
    "\n",
    "    print(p)\n",
    "\n",
    "\n",
    "    dir.create(dirname(output_plot_filepath), recursive = TRUE, showWarnings = FALSE)\n",
    "\n",
    "    ggsave(output_plot_filepath, plot = p, dpi = 500, bg = \"white\", width = 2100, height = 2100, units = \"px\")\n",
    "\n",
    "    group1_df <- filter(long_df, variable == \"value_group1\")\n",
    "    group2_df <- filter(long_df, variable == \"value_group2\")\n",
    "\n",
    "    if (set$ideal_value == 0) {\n",
    "        margin_adjustment <- 1 + margin_correct\n",
    "    } else {\n",
    "        margin_adjustment <- 1 - margin_correct\n",
    "    }\n",
    "\n",
    "    x_intersect_defaults_group1 <- find_intersection(group1_df, margin_adjustment * set$baseline_value_seurat_vs_scanpy_default)\n",
    "    x_intersect_defaults_group2 <- find_intersection(group2_df, margin_adjustment * set$baseline_value_seurat_vs_scanpy_default)\n",
    "\n",
    "    x_intersect_same_input_group1 <- find_intersection(group1_df, margin_adjustment * set$baseline_value_seurat_vs_scanpy_same_input)\n",
    "    x_intersect_same_input_group2 <- find_intersection(group2_df, margin_adjustment * set$baseline_value_seurat_vs_scanpy_same_input)\n",
    "\n",
    "\n",
    "    sink(file = output_stat_filepath, append = TRUE)\n",
    "    if (!(set$statistic_name %in% c(\"Markers Jaccard\", \"logFC CCC\", \"Adj p-value flip rate\"))) {\n",
    "        print(glue(\"{set$statistic_name} - Group 1 (Seurat) fraction intersecting with seu vs scan defaults: {x_intersect_defaults_group1}\"))\n",
    "    }\n",
    "\n",
    "    print(glue(\"{set$statistic_name} - Group 1 (Seurat) fraction intersecting with seu vs scan same input: {x_intersect_same_input_group1}\"))\n",
    "\n",
    "    if (!(set$statistic_name %in% c(\"Markers Jaccard\", \"logFC CCC\", \"Adj p-value flip rate\"))) {\n",
    "        print(glue(\"{set$statistic_name} - Group 2 (Scanpy) fraction intersecting with seu vs scan defaults: {x_intersect_defaults_group2}\"))\n",
    "    }\n",
    "\n",
    "    print(glue(\"{set$statistic_name} - Group 2 (Scanpy) fraction intersecting with seu vs scan same input: {x_intersect_same_input_group2}\"))\n",
    "    sink()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71553a49",
   "metadata": {},
   "source": [
    "WORK IN PROGRESS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbacd998",
   "metadata": {},
   "outputs": [],
   "source": [
    "if (length(downsampled_seeds_list) > 1) {\n",
    "    seed_intersections <- list(\n",
    "        Seurat = data.frame(matrix(ncol = length(downsampled_seeds_list), nrow = length(statistic_names_total))),\n",
    "        Scanpy = data.frame(matrix(ncol = length(downsampled_seeds_list), nrow = length(statistic_names_total)))\n",
    "    )\n",
    "    \n",
    "    # Set row and column names for each dataframe in the list\n",
    "    names(seed_intersections$Seurat) <- downsampled_seeds_list\n",
    "    rownames(seed_intersections$Seurat) <- statistic_names_total\n",
    "    \n",
    "    names(seed_intersections$Scanpy) <- downsampled_seeds_list\n",
    "    rownames(seed_intersections$Scanpy) <- statistic_names_total\n",
    "    \n",
    "    \n",
    "    \n",
    "    for (set in variable_sets) {\n",
    "        for (seed in downsampled_seeds_list) {\n",
    "            results_df <- data.frame(frac = numeric(), value_group1 = numeric(), value_group2 = numeric(), seed = numeric())\n",
    "    \n",
    "            for (package in c(\"seu\", \"scan\")) {\n",
    "                for (frac in frac_list) {\n",
    "                    \n",
    "                    frac_str <- gsub(\"\\\\.\", \"_\", as.character(frac)) # fraction of reads after downsampling, as string\n",
    "        \n",
    "                    stats_path <- get_stats_filepath(set, frac_str = frac_str, package = package, type_downsampled = type_downsampled, seed = seed)\n",
    "        \n",
    "                    # Read the file lines\n",
    "                    lines <- readLines(stats_path)\n",
    "        \n",
    "                    value_line <- grep(set$extraction_phrase, lines, value = TRUE)\n",
    "        \n",
    "                    # Extract the numerical variance using string manipulation\n",
    "                    value <- as.numeric(str_extract(value_line, \"(?<=: )\\\\d+(\\\\.\\\\d+)?(e[-+]?\\\\d+)?\"))\n",
    "        \n",
    "                    if (package == \"seu\") {\n",
    "                        new_row <- data.frame(frac = frac, value_group1 = value, value_group2 = NA, seed = seed)\n",
    "                    } else if (package == \"scan\") {\n",
    "                        new_row <- data.frame(frac = frac, value_group1 = NA, value_group2 = value, seed = seed)\n",
    "                    }\n",
    "                    # Add the results to the data frame\n",
    "                    results_df <- rbind(results_df, new_row)\n",
    "                }\n",
    "            }\n",
    "        \n",
    "            results_df2 <- results_df %>%\n",
    "                distinct() %>%\n",
    "                group_by(frac) %>%\n",
    "                summarize(\n",
    "                    value_group1 = max(value_group1, na.rm = TRUE),\n",
    "                    value_group2 = max(value_group2, na.rm = TRUE)\n",
    "                )\n",
    "            \n",
    "            long_df <- results_df2 %>%\n",
    "                pivot_longer(\n",
    "                    cols = c(value_group1, value_group2),\n",
    "                    names_to = \"variable\",\n",
    "                    values_to = \"value\"\n",
    "                )\n",
    "        \n",
    "            baseline_rows <- tibble(\n",
    "                frac = c(1.0, 1.0),\n",
    "                variable = c(\"value_group1\", \"value_group2\"),\n",
    "                value = c(ifelse(set$seurat_noise_value == 0, set$ideal_value, set$seurat_noise_value), ifelse(set$scanpy_noise_value == 0, set$ideal_value, set$scanpy_noise_value))\n",
    "            )\n",
    "        \n",
    "            long_df <- bind_rows(long_df, baseline_rows)\n",
    "        \n",
    "            group1_df <- filter(long_df, variable == \"value_group1\")\n",
    "            group2_df <- filter(long_df, variable == \"value_group2\")\n",
    "        \n",
    "            if (set$ideal_value == 0) {\n",
    "                margin_adjustment <- 1 + margin_correct\n",
    "            } else {\n",
    "                margin_adjustment <- 1 - margin_correct\n",
    "            }\n",
    "        \n",
    "            x_intersect_defaults_group1 <- find_intersection(group1_df, margin_adjustment * set$baseline_value_seurat_vs_scanpy_default)\n",
    "            x_intersect_defaults_group2 <- find_intersection(group2_df, margin_adjustment * set$baseline_value_seurat_vs_scanpy_default)\n",
    "        \n",
    "            x_intersect_same_input_group1 <- find_intersection(group1_df, margin_adjustment * set$baseline_value_seurat_vs_scanpy_same_input)\n",
    "            x_intersect_same_input_group2 <- find_intersection(group2_df, margin_adjustment * set$baseline_value_seurat_vs_scanpy_same_input)\n",
    "        \n",
    "            if (!(set$statistic_name %in% c(\"Markers Jaccard\", \"logFC CCC\", \"Adj p-value flip rate\"))) {\n",
    "                seed_intersections[[\"Seurat\"]][set$statistic_name, as.character(seed)] <- x_intersect_defaults_group1\n",
    "                seed_intersections[[\"Scanpy\"]][set$statistic_name, as.character(seed)] <- x_intersect_defaults_group2\n",
    "            } else {\n",
    "                seed_intersections[[\"Seurat\"]][set$statistic_name, as.character(seed)] <- x_intersect_same_input_group1\n",
    "                seed_intersections[[\"Scanpy\"]][set$statistic_name, as.character(seed)] <- x_intersect_same_input_group2\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    rownames(seed_intersections$Seurat)[rownames(seed_intersections$Seurat) == \"Median log SNN degree ratio\"] <- \"Median log(SNN degree ratio)\"\n",
    "    rownames(seed_intersections$Scanpy)[rownames(seed_intersections$Scanpy) == \"Median log SNN degree ratio\"] <- \"Median log(SNN degree ratio)\"\n",
    "    \n",
    "    seed_intersections$Seurat$Category <- factor(rownames(seed_intersections$Seurat), levels = rownames(seed_intersections$Seurat))\n",
    "    seed_intersections$Scanpy$Category <- factor(rownames(seed_intersections$Scanpy), levels = rownames(seed_intersections$Scanpy))\n",
    "    \n",
    "    df_long_seurat <- pivot_longer(seed_intersections$Seurat, cols = -Category, names_to = \"Sample\", values_to = \"Value\")\n",
    "    df_long_scanpy <- pivot_longer(seed_intersections$Scanpy, cols = -Category, names_to = \"Sample\", values_to = \"Value\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "535b460d",
   "metadata": {},
   "source": [
    "Bar charts summarizing all line graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0894a9b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "bar_chart_df <- data.frame(\n",
    "    Category = character(),\n",
    "    Seurat = numeric(),\n",
    "    Scanpy = numeric()\n",
    ")\n",
    "\n",
    "# bar_chart_df <- data.frame(\n",
    "#     category = c(\"Cell Filtering\", \"Gene Filtering\", \"HVG selection\", \"PCA\", \"KNN\", \"Clustering\", \"UMAP\", \"DE (marker genes)\", \"DE (markers)\", \"DE (logFC)\", \"DE (p-val adjusted)\"),\n",
    "#     Seurat = c(1, .4, .4, .4, .4, .4, .4, .4, .4, .4, .4),\n",
    "#     Scanpy = c(1, .4, .6, .3, .4, .4, .4, .4, .4, .4, .4))\n",
    "\n",
    "if (margin_correct == 0) {\n",
    "    aggregate_bar_chart_name <- \"bar_plot_across_reads.tiff\"\n",
    "} else {\n",
    "    aggregate_bar_chart_name <- glue(\"bar_plot_across_reads_{margin_correct}.tiff\")\n",
    "}\n",
    "\n",
    "# if (type_downsampled == \"cell\") {\n",
    "#     statistic_names_total <- statistic_names_total[2:8]\n",
    "# }\n",
    "\n",
    "lines <- readLines(output_stat_filepath)\n",
    "\n",
    "for (statistic in statistic_names_total) {\n",
    "    if (statistic %in% c(\"Markers Jaccard\", \"logFC CCC\", \"Adj p-value flip rate\")) {\n",
    "        extraction_phrase_ending <- \"same input\"\n",
    "    } else {\n",
    "        extraction_phrase_ending <- \"defaults\"\n",
    "    }\n",
    "\n",
    "    extraction_phrase_group1 <- glue(\"{statistic} - Group 1 \\\\(Seurat\\\\) fraction intersecting with seu vs scan {extraction_phrase_ending}: \")\n",
    "    value_line_group1 <- grep(extraction_phrase_group1, lines, value = TRUE)\n",
    "    value_group1 <- as.numeric(str_extract(value_line_group1, \"(?<=: )\\\\d+(\\\\.\\\\d+)?\"))\n",
    "\n",
    "    if (length(value_group1) == 0 || is.na(value_group1)) {\n",
    "        value_group1 <- 5e-3\n",
    "    }\n",
    "\n",
    "    extraction_phrase_group2 <- glue(\"{statistic} - Group 2 \\\\(Scanpy\\\\) fraction intersecting with seu vs scan {extraction_phrase_ending}: \")\n",
    "    value_line_group2 <- grep(extraction_phrase_group2, lines, value = TRUE)\n",
    "    value_group2 <- as.numeric(str_extract(value_line_group2, \"(?<=: )\\\\d+(\\\\.\\\\d+)?\"))\n",
    "\n",
    "    if (length(value_group2) == 0 || is.na(value_group2)) {\n",
    "        value_group2 <- 5e-3\n",
    "    }\n",
    "\n",
    "    new_row <- data.frame(Category = statistic, Seurat = value_group1, Scanpy = value_group2)\n",
    "\n",
    "    bar_chart_df <- rbind(bar_chart_df, new_row)\n",
    "}\n",
    "\n",
    "bar_chart_df$Category <- gsub(\"log SNN degree ratio\", \"log(SNN degree ratio)\", bar_chart_df$Category)\n",
    "\n",
    "bar_chart_df$Category <- factor(bar_chart_df$Category, levels = bar_chart_df$Category)\n",
    "\n",
    "long_bar_chart <- bar_chart_df %>%\n",
    "    pivot_longer(\n",
    "        cols = c(Seurat, Scanpy),\n",
    "        names_to = \"variable\",\n",
    "        values_to = \"value\"\n",
    "    )\n",
    "\n",
    "long_bar_chart$value <- ifelse(long_bar_chart$value < min(frac_list), min(frac_list), long_bar_chart$value)\n",
    "\n",
    "long_bar_chart$variable <- factor(long_bar_chart$variable, levels = c(\"Seurat\", \"Scanpy\"))\n",
    "\n",
    "\n",
    "\n",
    "aggregate_bar_chart <- ggplot(long_bar_chart, aes(x = Category, y = value, fill = variable)) +\n",
    "    geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "    # geom_text(aes(label = round(value, 2)),\n",
    "    #           position = position_dodge(width = 0.9),   # Adjust the position to align with the bars\n",
    "    #           vjust = -0.5,                             # Negative value for vertical adjustment to move text above bars\n",
    "    #           color = \"black\",                          # Text color\n",
    "    #           size = 3) +\n",
    "    scale_fill_manual(values = c(\"Seurat\" = \"#D55E00\", \"Scanpy\" = \"#56B4E9\")) +\n",
    "    labs(x = \"\", y = glue(\"Fraction of {tools::toTitleCase(type_downsampled)}s Needed\"), fill = \"Package\") +\n",
    "    coord_cartesian(ylim = c(0, 1)) +\n",
    "    theme_minimal() +\n",
    "    guides(fill = guide_legend(title = NULL)) +\n",
    "    theme(\n",
    "        text = element_text(family = \"Arial\"),\n",
    "        # legend.position = \"none\",\n",
    "        legend.text = element_text(size = rel(1.45)),\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1, size = rel(1.2)),\n",
    "        axis.title = element_text(size = rel(1.45))\n",
    "    ) # Rotate x-axis labels if needed\n",
    "\n",
    "if (length(downsampled_seeds_list) > 1) {\n",
    "    aggregate_bar_chart <- aggregate_bar_chart + \n",
    "        geom_point(data = df_long_seurat, \n",
    "                 aes(x = Category, y = Value, group = Category), \n",
    "                 inherit.aes = FALSE,\n",
    "                 position = position_nudge(x = -0.222),  # Adjust width to align with bars\n",
    "                 color = \"black\",  # Choose color to distinguish points\n",
    "                 fill = \"#D55E00\",\n",
    "                 shape = 21,\n",
    "                 size = 1.5, \n",
    "                 alpha = 0.5) +\n",
    "        geom_point(data = df_long_scanpy, \n",
    "                 aes(x = Category, y = Value, group = Category), \n",
    "                 inherit.aes = FALSE,\n",
    "                 position = position_nudge(x = 0.222),  # Adjust width to align with bars\n",
    "                 color = \"black\",  # Choose color to distinguish points\n",
    "                 fill = \"#56B4E9\",\n",
    "                 shape = 21,\n",
    "                 size = 1.5, \n",
    "                 alpha = 0.5)\n",
    "}\n",
    "\n",
    "print(aggregate_bar_chart)\n",
    "\n",
    "aggregate_plot_filepath <- glue(\"{output_path_base}/{aggregate_bar_chart_name}\")\n",
    "dir.create(dirname(aggregate_plot_filepath), recursive = TRUE, showWarnings = FALSE)\n",
    "\n",
    "ggsave(aggregate_plot_filepath, plot = aggregate_bar_chart, dpi = 500, bg = \"white\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a53409b8",
   "metadata": {},
   "source": [
    "Create the legend for all plots with dummy data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0028312c",
   "metadata": {},
   "outputs": [],
   "source": [
    "x <- 1:10\n",
    "df <- data.frame(\n",
    "    x = rep(x, 3),\n",
    "    y = c(x * 1.2, x * 0.8, x * 1.1),\n",
    "    group = factor(rep(c(\"Seurat Downsampled\", \"Scanpy Downsampled\", \"Seurat vs. Scanpy\"), each = 10), levels = c(\"Seurat Downsampled\", \"Scanpy Downsampled\", \"Seurat vs. Scanpy\")),\n",
    "    color = rep(c(\"#D55E00\", \"#56B4E9\", NA), each = 10),\n",
    "    linetype = rep(c(\"solid\", \"solid\", \"dashed\"), each = 10)\n",
    ")\n",
    "\n",
    "# Plot\n",
    "custom_legend <- ggplot(df, aes(x, y, group = group, color = group, linetype = group)) +\n",
    "    geom_line(linewidth = 0.5) +\n",
    "    scale_color_manual(values = c(\"Seurat Downsampled\" = \"#D55E00\", \"Scanpy Downsampled\" = \"#56B4E9\", \"Seurat vs. Scanpy\" = \"black\")) +\n",
    "    scale_linetype_manual(values = c(\"Seurat Downsampled\" = \"solid\", \"Scanpy Downsampled\" = \"solid\", \"Seurat vs. Scanpy (defaults)\" = \"dashed\")) +\n",
    "    theme(\n",
    "        text = element_text(family = \"Arial\"),\n",
    "        legend.title = element_blank(),\n",
    "        legend.text = element_text(size = 12)\n",
    "    ) +\n",
    "    guides(color = guide_legend(override.aes = list(linetype = c(\"solid\", \"solid\", \"dashed\"))))\n",
    "\n",
    "custom_legend\n",
    "\n",
    "custom_legend_filepath <- glue(\"{output_path_base}/line_plot_legend.tiff\")\n",
    "\n",
    "ggsave(custom_legend_filepath, plot = custom_legend, dpi = 500, bg = \"white\")"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
