{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "18baad57",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"Scanpy_v_Scanpy\"\n",
    "output: html_document\n",
    "date: \"2024-01-01\"\n",
    "---\n",
    "\n",
    "Select yaml file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5055b500",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml_file <- \"Fig3_scanpy\"  # Fig3_scanpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20e26602",
   "metadata": {},
   "outputs": [],
   "source": [
    "if (!requireNamespace(\"reticulate\", quietly = TRUE)) remotes::install_version(\"reticulate\", version = \"1.34.0\", upgrade = \"never\")\n",
    "\n",
    "using_colab <- reticulate::py_run_string(\"\n",
    "try:\n",
    "    import google.colab\n",
    "    using_colab = True\n",
    "except ImportError:\n",
    "    using_colab = False\n",
    "using_colab\n",
    "\")$using_colab\n",
    "\n",
    "if (using_colab) {\n",
    "    system(\"git clone https://github.com/josephrich98/scrnaseq_packages_and_versioning.git\", intern = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deade9ba",
   "metadata": {},
   "source": [
    "Load contents of yaml file into global R environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ff032c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml_dir <- glue::glue(\"{dirname(getwd())}/yaml\")\n",
    "yaml_file_path <- glue::glue(\"{yaml_dir}/{yaml_file}.yaml\")\n",
    "\n",
    "source(glue::glue(\"{dirname(getwd())}/scripts/load_yaml_contents.R\"))\n",
    "load_yaml_contents(yaml_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a6c170a",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_version_for_download <- gsub(\"_\", \".\", seurat_version)\n",
    "scanpy_version_for_download <- gsub(\"_\", \".\", scanpy_version)\n",
    "\n",
    "if (using_colab) {\n",
    "    py_command <- sprintf(\"import subprocess; subprocess.run(['pip', 'install', 'scanpy==%s', 'python-igraph==0.10.8', 'leidenalg==0.10.1', 'anndata==0.10.2', 'hdf5plugin==4.2.0', 'kb-python==0.27.3', 'umap-learn==0.5.2', 'louvain==0.8.1', 'git+https://github.com/has2k1/scikit-misc.git@269f61e'])\", scanpy_version_for_download)\n",
    "    \n",
    "    reticulate::py_run_string(py_command)\n",
    "}\n",
    "\n",
    "if (!requireNamespace(\"remotes\", quietly = TRUE)) install.packages(\"remotes\")\n",
    "\n",
    "if (!requireNamespace(\"tidyverse\", quietly = TRUE)) remotes::install_version(\"tidyverse\", version = \"2.0.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"rmarkdown\", quietly = TRUE)) remotes::install_version(\"rmarkdown\", version = \"2.25\", upgrade = \"never\")\n",
    "\n",
    "if (!requireNamespace(\"igraph\", quietly = TRUE)) pak::pak(\"igraph/rigraph\")\n",
    "if (!requireNamespace(\"Seurat\", quietly = TRUE)) remotes::install_version(\"Seurat\", version = seurat_version_for_download, upgrade = \"never\")\n",
    "if (!requireNamespace(\"Matrix\", quietly = TRUE)) remotes::install_version(\"Matrix\", version = \"1.6.4\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"patchwork\", quietly = TRUE)) remotes::install_version(\"patchwork\", version = \"1.1.3\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"eulerr\", quietly = TRUE)) remotes::install_version(\"eulerr\", version = \"7.0.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"scattermore\", quietly = TRUE)) remotes::install_version(\"scattermore\", version = \"1.2\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"assertthat\", quietly = TRUE)) remotes::install_version(\"assertthat\", version = \"0.2.1\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"pheatmap\", quietly = TRUE)) remotes::install_version(\"pheatmap\", version = \"1.0.12\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggforce\", quietly = TRUE)) remotes::install_version(\"ggforce\", version = \"0.4.1\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggplotify\", quietly = TRUE)) remotes::install_version(\"ggplotify\", version = \"0.1.2\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"mclust\", quietly = TRUE)) remotes::install_version(\"mclust\", version = \"6.0.1\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggalluvial\", quietly = TRUE)) remotes::install_version(\"ggalluvial\", version = \"0.12.5\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"UpSetR\", quietly = TRUE)) remotes::install_version(\"UpSetR\", version = \"1.4.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggpointdensity\", quietly = TRUE)) remotes::install_version(\"ggpointdensity\", version = \"0.1.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"dbscan\", quietly = TRUE)) remotes::install_version(\"dbscan\", version = \"1.1.12\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"presto\", quietly = TRUE)) remotes::install_github(\"immunogenomics/presto@31dc97f\", upgrade = \"never\")\n",
    "\n",
    "\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE)) remotes::install_version(\"BiocManager\", version = \"1.30.22\", upgrade = \"never\")\n",
    "bioconductor_version <- \"3.18\"\n",
    "\n",
    "if (!requireNamespace(\"BUSpaRse\", quietly = TRUE)) BiocManager::install(\"BUSpaRse\", version = bioconductor_version, update = FALSE)\n",
    "if (!requireNamespace(\"DropletUtils\", quietly = TRUE)) BiocManager::install(\"DropletUtils\", version = bioconductor_version, update = FALSE)\n",
    "if (!requireNamespace(\"biomaRt\", quietly = TRUE)) BiocManager::install(\"biomaRt\", version = bioconductor_version, update = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f450250",
   "metadata": {},
   "outputs": [],
   "source": [
    "if (scan1_all_genes_data_path == \"\") {\n",
    "    scan1_all_genes_data_path <- glue::glue(\"{data_path_root}/scanpyv1_9/adata_all_genes.h5ad\")\n",
    "}\n",
    "\n",
    "if (scan2_all_genes_data_path == \"\") {\n",
    "    scan2_all_genes_data_path <- glue::glue(\"{data_path_root}/scanpyv1_4/adata_all_genes.h5ad\")\n",
    "}\n",
    "\n",
    "if (scan1_data_path == \"\") {\n",
    "    scan1_data_path <- glue::glue(\"{data_path_root}/scanpyv1_9/adata.h5ad\")\n",
    "}\n",
    "\n",
    "if (scan2_data_path == \"\") {\n",
    "    scan2_data_path <- glue::glue(\"{data_path_root}/scanpyv1_4/adata.h5ad\")\n",
    "}\n",
    "\n",
    "if (scan1_markers_data_path == \"\") {\n",
    "    scan1_markers_data_path <- glue::glue(\"{data_path_root}/scanpyv1_9/markers.rds\")\n",
    "}\n",
    "\n",
    "if (scan2_markers_data_path == \"\") {\n",
    "    scan2_markers_data_path <- glue::glue(\"{data_path_root}/scanpyv1_4/markers.rds\")\n",
    "}\n",
    "\n",
    "if (output_base_path == \"\") {\n",
    "    output_base_path <- glue::glue(\"{project_base_path}/output/{data_name}/scanpyv{scanpy1_version}vs_scanpyv{scanpy2_version}\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cd6cd76",
   "metadata": {},
   "source": [
    "R Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "279aed3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda_env <- \"analysis_env\"\n",
    "Sys.setenv(RETICULATE_PYTHON = paste(\"/home/rstudio/.conda/envs\", conda_env, \"bin/python3.9\", sep = \"/\"))\n",
    "library(reticulate)\n",
    "use_condaenv(conda_env)\n",
    "library(Seurat)\n",
    "library(Matrix)\n",
    "library(tidyverse)\n",
    "library(patchwork)\n",
    "library(eulerr)\n",
    "library(scattermore)\n",
    "library(DropletUtils)\n",
    "library(glue)\n",
    "library(bluster)\n",
    "library(ggforce)\n",
    "library(ggplotify)\n",
    "library(grid)\n",
    "library(gtable)\n",
    "library(ggalluvial)\n",
    "theme_set(theme_bw())\n",
    "\n",
    "source(glue(\"{project_base_path}/scripts/data_analysis_helper.R\"))\n",
    "source(glue(\"{project_base_path}/scripts/plotting_and_stats.R\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fea0849",
   "metadata": {},
   "outputs": [],
   "source": [
    "scanpy1_minor_version <- as.integer(strsplit(scanpy1_version, split = \"\\\\.\")[[1]][2])\n",
    "scanpy2_minor_version <- as.integer(strsplit(scanpy2_version, split = \"\\\\.\")[[1]][2])\n",
    "\n",
    "scanpy_group_names <- list(Scanpy1 = scan1_name, Scanpy2 = scan2_name)\n",
    "\n",
    "group1_color <- \"#009E73\"\n",
    "group2_color <- \"#CC79A7\"\n",
    "\n",
    "# FALSE to have no save\n",
    "file_paths <- list(\n",
    "    filter_arguments = glue::glue(\"{output_base_path}/stats/filter_stats.txt\"),\n",
    "    euler_stats_before_QC_file = FALSE, # glue::glue(\"{output_base_path}/stats/euler_stats_beforeQC.txt\"),\n",
    "    euler_stats_after_QC_file = glue::glue(\"{output_base_path}/stats/euler_stats_afterQC.txt\"),\n",
    "    pca_knn_clustering_umap_file = glue::glue(\"{output_base_path}/stats/pca_knn_clustering_umap_stats.txt\"),\n",
    "    de_stats_file = glue::glue(\"{output_base_path}/stats/de_stats.txt\"),\n",
    "    pre_filtering_upset_cell = FALSE, # glue::glue(\"{output_base_path}/plots/pre_filtering_upset_cell.tiff\"),\n",
    "    pre_filtering_upset_gene = FALSE, # glue::glue(\"{output_base_path}/plots/pre_filtering_upset_gene.tiff\"),\n",
    "\n",
    "    knee_plot1 = FALSE, # glue::glue(\"{output_base_path}/plots/knee_plot_{scan1_name}.tiff\"),\n",
    "    knee_plot2 = FALSE, # glue::glue(\"{output_base_path}/plots/knee_plot_{scan2_name}.tiff\"),\n",
    "    umi_scatterplot = glue::glue(\"{output_base_path}/plots/umi_scatterplot.tiff\"),\n",
    "    violin_counts_comparison = glue::glue(\"{output_base_path}/plots/violin_counts_comparison.tiff\"),\n",
    "    scan1_violin_file_path_genes = FALSE, # glue::glue(\"{output_base_path}/plots/scan_violin_plot_genes.tiff\"),\n",
    "    scan1_violin_file_path_counts = FALSE, # glue::glue(\"{output_base_path}/plots/scan_violin_plot_counts.tiff\"),\n",
    "    scan1_violin_file_path_mt = FALSE, # glue::glue(\"{output_base_path}/plots/scan_violin_plot_mt.tiff\"),\n",
    "    scan2_violin_file_path_genes = FALSE, # glue::glue(\"{output_base_path}/plots/scan_violin_plot_genes.tiff\"),\n",
    "    scan2_violin_file_path_counts = FALSE, # glue::glue(\"{output_base_path}/plots/scan_violin_plot_counts.tiff\"),\n",
    "    scan2_violin_file_path_mt = FALSE, # glue::glue(\"{output_base_path}/plots/scan_violin_plot_mt.tiff\"),\n",
    "\n",
    "    upset_cells = glue::glue(\"{output_base_path}/plots/upset_cells.tiff\"),\n",
    "    upset_genes = glue::glue(\"{output_base_path}/plots/upset_genes.tiff\"),\n",
    "    upset_hvgs = glue::glue(\"{output_base_path}/plots/upset_hvgs.tiff\"),\n",
    "    upset_markers_genes_only = glue::glue(\"{output_base_path}/plots/upset_marker_genes_only.tiff\"),\n",
    "    upset_markers = glue::glue(\"{output_base_path}/plots/upset_markers.tiff\"),\n",
    "    euler_before_qc_cell_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/euler_cells_beforeQC.tiff\"),\n",
    "    euler_before_qc_gene_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/euler_genes_beforeQC.tiff\"),\n",
    "\n",
    "    euler_after_qc_cell_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/euler_cells_afterQC.tiff\"),\n",
    "    euler_after_qc_gene_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/euler_genes_afterQC.tiff\"),\n",
    "    euler_after_qc_hvg_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/euler_hvgs_afterQC.tiff\"),\n",
    "    euler_after_qc_marker_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/euler_markers.tiff\"),\n",
    "    euler_after_qc_marker_manual_bonferroni_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/euler_markers_manual_bonferroni.tiff\"),\n",
    "    euler_after_qc_marker_genes_only = FALSE, # glue::glue(\"{output_base_path}/plots/euler_markers_genes.tiff\"),\n",
    "\n",
    "    pca_elbow_filepath_combined = FALSE, # glue::glue(\"{output_base_path}/plots/pca_elbow_combined.tiff\"),\n",
    "    pca_12_overlay_filepath = glue::glue(\"{output_base_path}/plots/pca_scatterplot_12.tiff\"),\n",
    "    pca_34_overlay_filepath = FALSE, # glue::glue(\"{output_base_path}/plots/pca_scatterplot_34.tiff\"),\n",
    "    pca_loading_diffs = FALSE, # glue::glue(\"{output_base_path}/plots/pc_loading_diffs.tiff\"),\n",
    "    pca_eigs_diff = FALSE, # glue::glue(\"{output_base_path}/plots/pc_eig_diff.tiff\"),\n",
    "    pca_cluster_filepath_scan1 = FALSE, # glue::glue(\"{output_base_path}/plots/pca_scatterplot_clusters_scan_{scan1_name}.tiff\"),\n",
    "    pca_cluster_filepath_scan2 = FALSE, # glue::glue(\"{output_base_path}/plots/pca_scatterplot_clusters_scan_{scan2_name}.tiff\"),\n",
    "    combined_pc_variance_loadings_plot = glue::glue(\"{output_base_path}/plots/combined_pc_variance_loadings_plot.tiff\"),\n",
    "    jaccards = FALSE, # glue::glue(\"{output_base_path}/plots/jaccards.tiff\"),\n",
    "    knn_scatterplot = FALSE, # glue::glue(\"{output_base_path}/plots/knn_scatterplot.tiff\"),\n",
    "    jaccard_degree_scatterplot = glue::glue(\"{output_base_path}/plots/jaccard_degree_scatterplot.tiff\"),\n",
    "    pheatmap = FALSE, # glue::glue(\"{output_base_path}/plots/cluster_pheatmap.tiff\"),\n",
    "    alluvial = glue::glue(\"{output_base_path}/plots/cluster_alluvial.tiff\"),\n",
    "    alluvial_legend = glue::glue(\"{output_base_path}/plots/cluster_alluvial_legend.tiff\"),\n",
    "    alluvial_legend_high_alpha = glue::glue(\"{output_base_path}/plots/cluster_alluvial_legend_high_alpha.tiff\"),\n",
    "    umap_scan1 = glue::glue(\"{output_base_path}/plots/umap_scan_{scan1_name}.tiff\"),\n",
    "    umap_scan2 = glue::glue(\"{output_base_path}/plots/umap_scan_{scan2_name}.tiff\"),\n",
    "    umap_scan1_clusters_scan2 = glue::glue(\"{output_base_path}/plots/umap_scan_{scan1_name}_clusters_{scan2_name}.tiff\"),\n",
    "    umap_scan2_clusters_scan1 = glue::glue(\"{output_base_path}/plots/umap_scan_{scan2_name}_clusters_{scan1_name}.tiff\"),\n",
    "    umap_jaccard_degree_scatterplot = glue::glue(\"{output_base_path}/plots/umap_jaccard_degree_scatterplot.tiff\"),\n",
    "    umap_jaccard_knn_density = glue::glue(\"{output_base_path}/plots/umap_jaccard_knn_density.tiff\"),\n",
    "    umap_jaccard_knn_density_scan1_facet = glue::glue(\"{output_base_path}/plots/umap_jaccard_knn_density_scan1_facet.tiff\"),\n",
    "    umap_jaccard_knn_density_scan2_facet = glue::glue(\"{output_base_path}/plots/umap_jaccard_knn_density_scan2_facet.tiff\"),\n",
    "    umap_alluvial = glue::glue(\"{output_base_path}/plots/umap_alluvial.tiff\"),\n",
    "    umap_alluvial_legend = glue::glue(\"{output_base_path}/plots/umap_alluvial_legend.tiff\"),\n",
    "    umap_umap_leiden_scan1 = glue::glue(\"{output_base_path}/plots/umap_umap_leiden_scan1.tiff\"),\n",
    "    umap_umap_leiden_scan2 = glue::glue(\"{output_base_path}/plots/umap_umap_leiden_scan2.tiff\"),\n",
    "    logFC_histogram_magnitude_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/logFC_histogram_magnitude.tiff\"),\n",
    "    logFC_histogram_signed_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/logFC_histogram_signed.tiff\"),\n",
    "    wilcoxon_histogram_magnitude_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/wilcoxon_histogram_magnitude.tiff\"),\n",
    "    wilcoxon_histogram_signed_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/wilcoxon_histogram_signed.tiff\"),\n",
    "\n",
    "    logFC_scatterplot_file_path = glue::glue(\"{output_base_path}/plots/logFC_scatterplot.tiff\"),\n",
    "    wilcoxon_scatterplot_file_path = glue::glue(\"{output_base_path}/plots/wilcoxon_scatterplot.tiff\"),\n",
    "    logFC_scatterplot_file_path_with_legend = glue::glue(\"{output_base_path}/plots/logFC_scatterplot_with_legend.tiff\"),\n",
    "    logFC_scatterplot_outliers_removed_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/logFC_scatterplot_no_outliers.tiff\"),\n",
    "    wilcoxon_scatterplot_outliers_removed_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/wilcoxon_scatterplot_no_outliers.tiff\"),\n",
    "\n",
    "    logFC_boxplot_magnitude_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/logFC_boxplot_magnitude.tiff\"),\n",
    "    logFC_boxplot_signed_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/logFC_boxplot_signed.tiff\"),\n",
    "    wilcoxon_boxplot_magnitude_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/wilcoxon_boxplot_magnitude.tiff\"),\n",
    "    wilcoxon_boxplot_signed_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/wilcoxon_boxplot_signed.tiff\"),\n",
    "\n",
    "    FC_histogram_magnitude_file_path = FALSE, # glue::glue(\"{output_base_path}/plots/FC_histogram_magnitude.tiff\"),\n",
    "    FC_histogram_signed_file_path = FALSE # glue::glue(\"{output_base_path}/plots/FC_histogram_signed.tiff\")\n",
    ")\n",
    "\n",
    "\n",
    "if (save_data) {\n",
    "    for (path in file_paths) {\n",
    "        if (is.character(path)) {\n",
    "            # Extract the directory part of the path\n",
    "            specific_output_path <- dirname(path)\n",
    "\n",
    "            # Create the directory if it does not exist\n",
    "            if (!dir.exists(specific_output_path)) {\n",
    "                dir.create(specific_output_path, recursive = TRUE, showWarnings = FALSE)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "} else {\n",
    "    for (i in seq_along(file_paths)) {\n",
    "        file_paths[[i]] <- FALSE\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4e1dca1",
   "metadata": {},
   "source": [
    "Download data if necessary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa43fe28",
   "metadata": {},
   "outputs": [],
   "source": [
    "py_run_string('import sys\n",
    "sys.path.append(f\"{r.project_base_path}/scripts\")\n",
    "from download_data import *\n",
    "\n",
    "if r.download_data:\n",
    "    download_and_extract(r.doi, r.data_name, r.data_path_root)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b0e3864",
   "metadata": {},
   "outputs": [],
   "source": [
    "scanpy_hvg_flavor <- \"seurat\"\n",
    "n_top_genes <- NULL\n",
    "scanpy_scale_max <- NULL\n",
    "scanpy_pca_zero_center <- TRUE\n",
    "scan_n_neighbors <- 15\n",
    "scanpy_clustering_algorithm <- \"leiden\"\n",
    "scanpy_resolution <- 1\n",
    "scanpy_cluster_iters <- -1\n",
    "scanpy_umap_min_dist <- 0.5\n",
    "scanpy_correction_method <- \"benjamini-hochberg\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a160407c",
   "metadata": {},
   "outputs": [],
   "source": [
    "py_run_string('import os \n",
    "import shutil\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import anndata\n",
    "import hdf5plugin')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b40949a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "py_run_string('adata1_all_genes = sc.read_h5ad(r.scan1_all_genes_data_path)\n",
    "adata2_all_genes = sc.read_h5ad(r.scan2_all_genes_data_path)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a667248",
   "metadata": {},
   "outputs": [],
   "source": [
    "upset_cell <- make_upset_scanpy(py$adata1_all_genes, py$adata2_all_genes, comparison = \"Cell\", group_names = scanpy_group_names, save = file_paths$upset_cells)\n",
    "upset_gene <- make_upset_scanpy(py$adata1_all_genes, py$adata2_all_genes, comparison = \"Gene\", group_names = scanpy_group_names, save = file_paths$upset_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3bcd2cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "py_run_string('adata1 = sc.read_h5ad(r.scan1_data_path)\n",
    "adata2 = sc.read_h5ad(r.scan2_data_path)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4a68e5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "upset_hvg <- make_upset_scanpy(py$adata1, py$adata2, comparison = \"HVG\", group_names = scanpy_group_names, save = file_paths$upset_hvgs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dea6f68e",
   "metadata": {},
   "outputs": [],
   "source": [
    "py_run_string('scanpy_highly_variable_genes_list1 = adata1.var[adata1.var[\\'highly_variable\\']].index.tolist()\n",
    "scanpy_highly_variable_genes_list2 = adata2.var[adata2.var[\\'highly_variable\\']].index.tolist()\n",
    "\n",
    "var_explained_py1 = adata1.uns[\\'pca\\'][\\'variance_ratio\\']\n",
    "var_explained_py2 = adata2.uns[\\'pca\\'][\\'variance_ratio\\']')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5b4c9a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "hvgs <- list(Scanpy1 = py$scanpy_highly_variable_genes_list1, Scanpy2 = py$scanpy_highly_variable_genes_list2)\n",
    "\n",
    "scan1_inds <- as.vector(py$adata1$obs_names$values)\n",
    "scan1_genes <- as.vector(py$adata1$var_names$values)\n",
    "\n",
    "scan2_inds <- as.vector(py$adata2$obs_names$values)\n",
    "scan2_genes <- as.vector(py$adata2$var_names$values)\n",
    "\n",
    "overlapping_inds <- intersect(scan1_inds, scan2_inds)\n",
    "overlapping_genes <- intersect(scan1_genes, scan2_genes)\n",
    "\n",
    "var_explained_py1 <- py$var_explained_py1\n",
    "var_explained_py2 <- py$var_explained_py2\n",
    "\n",
    "eigs_df <- tibble(\n",
    "    Scanpy1 = var_explained_py1,\n",
    "    Scanpy2 = var_explained_py2,\n",
    "    PC = 1:50\n",
    ")\n",
    "\n",
    "combined_pc_variance <- plot_var_explained(eigs_df, npcs = 50, group_names = unlist(scanpy_group_names), save = file_paths$pca_elbow_filepath_combined)\n",
    "combined_pc_variance\n",
    "\n",
    "pca_embeddings1 <- py$adata1$obsm[\"X_pca\"]\n",
    "pca_embeddings2 <- py$adata2$obsm[\"X_pca\"]\n",
    "\n",
    "rownames(pca_embeddings1) <- scan1_inds\n",
    "rownames(pca_embeddings2) <- scan2_inds\n",
    "\n",
    "all.equal(pca_embeddings1, pca_embeddings2)\n",
    "\n",
    "if (!identical(scan1_inds, scan2_inds)) {\n",
    "    pca_embeddings1 <- pca_embeddings1[rownames(pca_embeddings1) %in% overlapping_inds, ]\n",
    "    pca_embeddings2 <- pca_embeddings2[rownames(pca_embeddings2) %in% overlapping_inds, ]\n",
    "    \n",
    "    pca_embeddings2 <- pca_embeddings2[match(rownames(pca_embeddings1), rownames(pca_embeddings2)), ]\n",
    "}\n",
    "\n",
    "pca12_plot <- plot_pca_compare(pca_embeddings1, pca_embeddings2, group1_name = \"Scanpy1\", group2_name = \"Scanpy2\", group_labels = unlist(scanpy_group_names), save = file_paths$pca_12_overlay_filepath)\n",
    "pca12_plot\n",
    "\n",
    "is_hvg_py1 <- py$adata1$var$highly_variable\n",
    "is_hvg_py2 <- py$adata2$var$highly_variable\n",
    "\n",
    "pca_loadings_scan1 <- py$adata1$varm[\"PCs\"]\n",
    "pca_loadings_scan2 <- py$adata2$varm[\"PCs\"]\n",
    "\n",
    "pca_loadings_scan1 <- pca_loadings_scan1[is_hvg_py1, ]\n",
    "pca_loadings_scan2 <- pca_loadings_scan2[is_hvg_py2, ]\n",
    "\n",
    "rownames(pca_loadings_scan1) <- hvgs$Scanpy1\n",
    "rownames(pca_loadings_scan2) <- hvgs$Scanpy2\n",
    "\n",
    "df_loadings <- make_pc_diffs_df(list(\n",
    "    Scanpy1 = pca_loadings_scan1,\n",
    "    Scanpy2 = pca_loadings_scan2\n",
    "), npcs = 50)\n",
    "\n",
    "mean_loadings_diff <- mean(df_loadings$differences[1:3])\n",
    "\n",
    "mylist <- list(\n",
    "    Scanpy1 = pca_loadings_scan1,\n",
    "    Scanpy2 = pca_loadings_scan2\n",
    ")\n",
    "\n",
    "loading_diffs <- plot_loading_diffs(df_loadings, save = file_paths$pca_loading_diffs)\n",
    "\n",
    "loading_diffs\n",
    "\n",
    "combined_plot <- make_combined_pc_variance_loadings_plot(combined_pc_variance, loading_diffs, save = file_paths$combined_pc_variance_loadings_plot)\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33752152",
   "metadata": {},
   "outputs": [],
   "source": [
    "py_run_string('snn_graph_scan1 = adata1.obsp[\\'connectivities\\']\n",
    "snn_graph_scan2 = adata2.obsp[\\'connectivities\\']')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "012431e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "scan_snn_b1 <- py$snn_graph_scan1 > 0\n",
    "scan_snn_b2 <- py$snn_graph_scan2 > 0\n",
    "\n",
    "rownames(scan_snn_b1) <- as.vector(py$adata1$obs_names$values)\n",
    "colnames(scan_snn_b1) <- as.vector(py$adata1$obs_names$values)\n",
    "\n",
    "rownames(scan_snn_b2) <- as.vector(py$adata2$obs_names$values)\n",
    "colnames(scan_snn_b2) <- as.vector(py$adata2$obs_names$values)\n",
    "\n",
    "if (!identical(scan1_inds, scan2_inds)) {\n",
    "    scan_snn_b1 <- scan_snn_b1[overlapping_inds, overlapping_inds]\n",
    "    scan_snn_b2 <- scan_snn_b2[overlapping_inds, overlapping_inds]\n",
    "}\n",
    "\n",
    "scan1_list <- mat2list(scan_snn_b1)\n",
    "scan2_list <- mat2list(scan_snn_b2)\n",
    "\n",
    "jaccards <- find_jaccards(list(Scanpy1 = scan1_list, Scanpy2 = scan2_list))\n",
    "\n",
    "median_jaccard <- median(jaccards$Jaccard)\n",
    "\n",
    "nei_sizes <- tibble(\n",
    "    Scanpy1 = lengths(scan1_list),\n",
    "    Scanpy2 = lengths(scan2_list)\n",
    ")\n",
    "\n",
    "nei_pairs <- make_pairwise_df(nei_sizes)\n",
    "\n",
    "\n",
    "jaccards$degree_ratio <- nei_pairs$value1 / nei_pairs$value2\n",
    "jaccards$logged_degree_ratio <- log(jaccards$degree_ratio, base = 2)\n",
    "\n",
    "jaccards$logged_degree_ratio[jaccards$logged_degree_ratio == -Inf] <- -10\n",
    "jaccards$logged_degree_ratio[jaccards$logged_degree_ratio == Inf] <- 10\n",
    "\n",
    "jaccards$jaccard_logged <- log(jaccards$Jaccard, base = 2)\n",
    "\n",
    "median_magnitude_logged_degree_ratio <- median(abs(jaccards$logged_degree_ratio))\n",
    "\n",
    "jaccard_degree_scatterplot <- make_snn_jaccard_degree_scatterplot(jaccards, save = file_paths$jaccard_degree_scatterplot)\n",
    "jaccard_degree_scatterplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1892b235",
   "metadata": {},
   "outputs": [],
   "source": [
    "scan1_clusters <- py$adata1$obs[[scanpy_clustering_algorithm]]\n",
    "scan2_clusters <- py$adata2$obs[[scanpy_clustering_algorithm]]\n",
    "\n",
    "names(scan1_clusters) <- scan1_inds\n",
    "names(scan2_clusters) <- scan2_inds\n",
    "\n",
    "if (!identical(scan1_inds, scan2_inds)) {\n",
    "    scan1_clusters <- scan1_clusters[names(scan1_clusters) %in% overlapping_inds]\n",
    "    scan2_clusters <- scan2_clusters[names(scan2_clusters) %in% overlapping_inds]\n",
    "\n",
    "    cell_order <- names(scan1_clusters)\n",
    "    scan2_clusters <- scan2_clusters[match(cell_order, names(scan2_clusters))]\n",
    "}\n",
    "\n",
    "scan1_clusters_vector <- as.vector(scan1_clusters)\n",
    "scan2_clusters_vector <- as.vector(scan2_clusters)\n",
    "ari_value <- mclust::adjustedRandIndex(scan1_clusters_vector, scan2_clusters_vector)\n",
    "\n",
    "scan1_clusters <- factor(scan1_clusters, levels = sort(as.numeric(levels(scan1_clusters))))\n",
    "scan2_clusters <- factor(scan2_clusters, levels = sort(as.numeric(levels(scan2_clusters))))\n",
    "\n",
    "df <- tibble(\n",
    "    Scanpy1 = scan1_clusters,\n",
    "    Scanpy2 = scan2_clusters\n",
    ")\n",
    "\n",
    "df <- setNames(df, unlist(scanpy_group_names))\n",
    "\n",
    "clus_df_gather <- get_alluvial_df(df)\n",
    "\n",
    "clus_df_gather <- clus_df_gather %>% mutate(\n",
    "    group1_column_original_clusters := as.numeric(as.character(.data[[scan1_name]])),\n",
    "    group2_column_original_clusters := as.numeric(as.character(.data[[scan2_name]]))\n",
    ")\n",
    "\n",
    "clus_df_gather <- sort_clusters_by_agreement(clus_df_gather, stable_column = scan1_name, reordered_column = scan2_name)\n",
    "# clus_df_gather <- sort_clusters_by_agreement(clus_df_gather, stable_column = scan1_name, reordered_column = scan2_name)\n",
    "\n",
    "alluvial_plot <- plot_alluvial(clus_df_gather, color_boxes = TRUE, color_bands = FALSE, group1_name = scan1_name, group2_name = scan2_name, save = file_paths$alluvial)\n",
    "alluvial_plot_legend <- plot_alluvial(clus_df_gather, color_boxes = TRUE, color_bands = TRUE, alluvial_alpha = 0.5, group1_name = scan1_name, group2_name = scan2_name, save = file_paths$alluvial_legend)\n",
    "alluvial_plot_legend_high_alpha <- plot_alluvial(clus_df_gather, color_boxes = TRUE, color_bands = TRUE, alluvial_alpha = 1, group1_name = scan1_name, group2_name = scan2_name, save = file_paths$alluvial_legend_high_alpha)\n",
    "\n",
    "alluvial_plot\n",
    "alluvial_plot_legend\n",
    "\n",
    "unique_mapping <- clus_df_gather %>%\n",
    "    ungroup() %>%\n",
    "    select(!!sym(scan2_name), group2_column_original_clusters) %>%\n",
    "    distinct() %>%\n",
    "    arrange(group2_column_original_clusters)\n",
    "\n",
    "scanpy_clusters_df <- data.frame(cell_id = scan2_inds, group2_cluster = as.numeric(as.character(py$adata2$obs[[scanpy_clustering_algorithm]])))\n",
    "\n",
    "mapped_data <- scanpy_clusters_df %>% left_join(unique_mapping, by = c(\"group2_cluster\" = \"group2_column_original_clusters\"))\n",
    "\n",
    "mapped_data[[scan2_name]] <- as.character(mapped_data[[scan2_name]])\n",
    "\n",
    "named_vector <- setNames(mapped_data[[scan2_name]], mapped_data$cell_id)\n",
    "\n",
    "scan2_clusters_renumbered <- factor(named_vector)\n",
    "\n",
    "scan1_cluster_data_original <- py$adata1$obs[[scanpy_clustering_algorithm]]\n",
    "scan2_cluster_data_original <- py$adata2$obs[[scanpy_clustering_algorithm]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5f08e6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors_group2 <- find_group2_colors(clus_df_gather, scan1_name, scan2_name)\n",
    "\n",
    "umap_plots <- plot_umap(group1_umap_info = py$adata1$obsm[\"X_umap\"], group1_clusters = py$adata1$obs[scanpy_clustering_algorithm][, 1], group2_umap_info = py$adata2$obsm[\"X_umap\"], group2_clusters = scan2_clusters_renumbered, group1 = scan1_name, group2 = scan2_name, colors_group2 = colors_group2, save = c(file_paths$umap_scan1, file_paths$umap_scan2))\n",
    "scan1_umap <- umap_plots[[1]]\n",
    "scan2_umap <- umap_plots[[2]]\n",
    "\n",
    "scan1_umap\n",
    "scan2_umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2f759dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "result1 <- readRDS(scan1_markers_data_path)\n",
    "result2 <- readRDS(scan2_markers_data_path)\n",
    "\n",
    "scan1_filtered_markers <- result1 %>% filter(p_val_adj < 0.05)\n",
    "scan2_filtered_markers <- result2 %>% filter(p_val_adj < 0.05)\n",
    "\n",
    "vectorized_scan1_filtered_markers <- unique(scan1_filtered_markers$gene)\n",
    "vectorized_scan2_filtered_markers <- unique(scan2_filtered_markers$gene)\n",
    "\n",
    "\n",
    "markers_euler_genes_only <- make_euler_scanpy(vectorized_scan1_filtered_markers, vectorized_scan2_filtered_markers, comparison = \"Marker Gene\", group_names = scanpy_group_names, save_plot = file_paths$euler_after_qc_marker_genes_only, save_stats = file_paths$de_stats_file)\n",
    "markers_euler_genes_only\n",
    "\n",
    "upset_marker_gene_only <- make_upset_scanpy(vectorized_scan1_filtered_markers, vectorized_scan2_filtered_markers, comparison = \"Marker Gene\", group_names = scanpy_group_names, save = file_paths$upset_markers_genes_only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2c97701",
   "metadata": {},
   "outputs": [],
   "source": [
    "result1_same_clusters <- result1\n",
    "\n",
    "# result1_same_clusters <- result1_same_clusters %>%\n",
    "#     dplyr::rename(avg_log2FC = log_fc, p_val_adj = p_value_adj, p_val = p_value)\n",
    "\n",
    "if (!all(scan1_clusters == scan2_clusters)) {\n",
    "    py$adata2$obs[scanpy_clustering_algorithm] <- scan1_clusters\n",
    "    \n",
    "    adata2_input_object_path <- glue::glue(\"{output_base_path}/data_files/adata2_same_clusters.h5ad\")\n",
    "    \n",
    "    if (!dir.exists(dirname(adata2_input_object_path))) {\n",
    "        dir.create(dirname(adata2_input_object_path), recursive = TRUE, showWarnings = FALSE)\n",
    "    }\n",
    "    \n",
    "    py$adata2$write_h5ad(adata2_input_object_path, compression = py$hdf5plugin$FILTERS$zstd)\n",
    "    \n",
    "    adata2_output_object_path <- glue::glue(\"{output_base_path}/data_files/adata2_same_clusters_de.h5ad\")\n",
    "    \n",
    "    conda_environment_with_older_scanpy = \"sc14_umap5\"\n",
    "    python_script_path <- glue(\"{project_base_path}/scripts/run_scanpy14_rank_genes_groups.py\")\n",
    "    \n",
    "    conda_path <- paste(\"/home/rstudio/.conda/envs\", conda_environment_with_older_scanpy, \"bin/python3.9\", sep = \"/\")\n",
    "\n",
    "    cmd_adata <- sprintf(\"%s %s '%s' '%s' '%s' '%s'\", conda_path, python_script_path, adata2_input_object_path, adata2_output_object_path, scanpy_clustering_algorithm, scanpy_correction_method)\n",
    "    \n",
    "    system(cmd_adata)\n",
    "    \n",
    "    py$adata2_same_clusters <- py$sc$read_h5ad(adata2_output_object_path)\n",
    "    \n",
    "    scanpy_minor_version <- 4\n",
    "    result2_same_clusters <- get_py_de_results(\"adata2_same_clusters\")\n",
    "    \n",
    "    result2_same_clusters <- result2_same_clusters %>%\n",
    "        dplyr::rename(avg_log2FC = log_fc, p_val_adj = p_value_adj, p_val = p_value)\n",
    "} else {\n",
    "    result2_same_clusters <- result2\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "963e8acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "result1_markers_df <- result1_same_clusters %>% select(gene = gene, cluster = cluster)\n",
    "result2_markers_df <- result2_same_clusters %>% select(gene = gene, cluster = cluster)\n",
    "\n",
    "vectorized_scan1_markers <- paste(result1_markers_df$gene, result1_markers_df$cluster, sep = \"-\")\n",
    "vectorized_scan2_markers <- paste(result2_markers_df$gene, result2_markers_df$cluster, sep = \"-\")\n",
    "\n",
    "markers_euler <- make_euler_scanpy(vectorized_scan1_markers, vectorized_scan2_markers, comparison = \"Marker\", group_names = scanpy_group_names, save_plot = file_paths$euler_after_qc_marker_file_path, save_stats = file_paths$de_stats_file)\n",
    "markers_euler\n",
    "\n",
    "upset_markers_all <- make_upset_scanpy(vectorized_scan1_markers, vectorized_scan2_markers, comparison = \"Marker\", group_names = scanpy_group_names, save = file_paths$upset_markers)\n",
    "\n",
    "markers2 <- result1_same_clusters |>\n",
    "    inner_join(result2_same_clusters, by = c(\"cluster\", \"gene\"), suffix = c(glue(\".{scan1_name}\"), glue(\".{scan2_name}\")))\n",
    "\n",
    "markers2 <- markers2 |>\n",
    "    mutate(cluster = factor(cluster, levels = as.character(seq_len(length(unique(cluster))) - 1)))\n",
    "\n",
    "markers2 <- markers2 |>\n",
    "    group_by(cluster) |>\n",
    "    mutate(rank_r = seq_along(gene))\n",
    "\n",
    "\n",
    "markers2[[glue(\"FC.{scan1_name}\")]] <- 2^markers2[[glue(\"avg_log2FC.{scan1_name}\")]]\n",
    "markers2[[glue(\"FC.{scan2_name}\")]] <- 2^markers2[[glue(\"avg_log2FC.{scan2_name}\")]]\n",
    "\n",
    "markers2 <- calculate_de_stats(markers2, group1_name = scan1_name, group2_name = scan2_name, save = file_paths$de_stats_file)\n",
    "\n",
    "markers2[[glue(\"p_val_adj.{scan1_name}\")]][markers2[[glue(\"p_val_adj.{scan1_name}\")]] == 0] <- .Machine$double.xmin\n",
    "markers2[[glue(\"p_val_adj.{scan2_name}\")]][markers2[[glue(\"p_val_adj.{scan2_name}\")]] == 0] <- .Machine$double.xmin"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8ae6d71",
   "metadata": {},
   "source": [
    "Potentially run in scanpy_de_plots file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f489439c",
   "metadata": {},
   "outputs": [],
   "source": [
    "logFC_scatterplot <- plot_scatterplot_de_logfc(markers2, group1_name = scan1_name, group2_name = scan2_name, ccc = markers2$CCC[1], save = file_paths$logFC_scatterplot_file_path, outliers_excluded = FALSE)\n",
    "pvaladj_scatterplot <- plot_scatterplot_de_wilcoxon(markers2, group1_name = scan1_name, group2_name = scan2_name, save = file_paths$wilcoxon_scatterplot_file_path, outliers_excluded = FALSE)\n",
    "\n",
    "logFC_scatterplot_with_legend <- plot_scatterplot_de_logfc(markers2, group1_name = scan1_name, group2_name = scan2_name, ccc = markers2$CCC[1], save = file_paths$logFC_scatterplot_file_path_with_legend, outliers_excluded = FALSE, show_legend = TRUE)\n",
    "\n",
    "\n",
    "logFC_scatterplot\n",
    "logFC_scatterplot_with_legend\n",
    "\n",
    "pvaladj_scatterplot"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
