{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5b046885",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"Seurat version comparison\"\n",
    "output: html_document\n",
    "date: \"2023-11-15\"\n",
    "---\n",
    "\n",
    "Select yaml file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ea365fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml_file <- \"Fig3_seurat\"  # Fig3_seurat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cc68d47",
   "metadata": {},
   "outputs": [],
   "source": [
    "if (!requireNamespace(\"reticulate\", quietly = TRUE)) remotes::install_version(\"reticulate\", version = \"1.34.0\", upgrade = \"never\")\n",
    "\n",
    "using_colab <- reticulate::py_run_string(\"\n",
    "try:\n",
    "    import google.colab\n",
    "    using_colab = True\n",
    "except ImportError:\n",
    "    using_colab = False\n",
    "using_colab\n",
    "\")$using_colab\n",
    "\n",
    "if (using_colab) {\n",
    "    system(\"git clone https://github.com/josephrich98/scrnaseq_packages_and_versioning.git\", intern = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ba7c197",
   "metadata": {},
   "source": [
    "Load contents of yaml file into global R environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7d78c6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml_dir <- glue::glue(\"{dirname(getwd())}/yaml\")\n",
    "yaml_file_path <- glue::glue(\"{yaml_dir}/{yaml_file}.yaml\")\n",
    "\n",
    "source(glue::glue(\"{dirname(getwd())}/scripts/load_yaml_contents.R\"))\n",
    "load_yaml_contents(yaml_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01a06565",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_version_for_download <- gsub(\"_\", \".\", seurat_version)\n",
    "scanpy_version_for_download <- gsub(\"_\", \".\", scanpy_version)\n",
    "\n",
    "if (using_colab) {\n",
    "    py_command <- sprintf(\"import subprocess; subprocess.run(['pip', 'install', 'scanpy==%s', 'python-igraph==0.10.8', 'leidenalg==0.10.1', 'anndata==0.10.2', 'hdf5plugin==4.2.0', 'kb-python==0.27.3', 'umap-learn==0.5.2', 'louvain==0.8.1', 'git+https://github.com/has2k1/scikit-misc.git@269f61e'])\", scanpy_version_for_download)\n",
    "    \n",
    "    reticulate::py_run_string(py_command)\n",
    "}\n",
    "\n",
    "if (!requireNamespace(\"remotes\", quietly = TRUE)) install.packages(\"remotes\")\n",
    "\n",
    "if (!requireNamespace(\"tidyverse\", quietly = TRUE)) remotes::install_version(\"tidyverse\", version = \"2.0.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"rmarkdown\", quietly = TRUE)) remotes::install_version(\"rmarkdown\", version = \"2.25\", upgrade = \"never\")\n",
    "\n",
    "if (!requireNamespace(\"igraph\", quietly = TRUE)) pak::pak(\"igraph/rigraph\")\n",
    "if (!requireNamespace(\"Seurat\", quietly = TRUE)) remotes::install_version(\"Seurat\", version = seurat_version_for_download, upgrade = \"never\")\n",
    "if (!requireNamespace(\"Matrix\", quietly = TRUE)) remotes::install_version(\"Matrix\", version = \"1.6.4\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"patchwork\", quietly = TRUE)) remotes::install_version(\"patchwork\", version = \"1.1.3\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"eulerr\", quietly = TRUE)) remotes::install_version(\"eulerr\", version = \"7.0.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"scattermore\", quietly = TRUE)) remotes::install_version(\"scattermore\", version = \"1.2\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"assertthat\", quietly = TRUE)) remotes::install_version(\"assertthat\", version = \"0.2.1\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"pheatmap\", quietly = TRUE)) remotes::install_version(\"pheatmap\", version = \"1.0.12\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggforce\", quietly = TRUE)) remotes::install_version(\"ggforce\", version = \"0.4.1\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggplotify\", quietly = TRUE)) remotes::install_version(\"ggplotify\", version = \"0.1.2\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"mclust\", quietly = TRUE)) remotes::install_version(\"mclust\", version = \"6.0.1\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggalluvial\", quietly = TRUE)) remotes::install_version(\"ggalluvial\", version = \"0.12.5\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"UpSetR\", quietly = TRUE)) remotes::install_version(\"UpSetR\", version = \"1.4.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"ggpointdensity\", quietly = TRUE)) remotes::install_version(\"ggpointdensity\", version = \"0.1.0\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"dbscan\", quietly = TRUE)) remotes::install_version(\"dbscan\", version = \"1.1.12\", upgrade = \"never\")\n",
    "if (!requireNamespace(\"presto\", quietly = TRUE)) remotes::install_github(\"immunogenomics/presto@31dc97f\", upgrade = \"never\")\n",
    "\n",
    "\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE)) remotes::install_version(\"BiocManager\", version = \"1.30.22\", upgrade = \"never\")\n",
    "bioconductor_version <- \"3.18\"\n",
    "\n",
    "if (!requireNamespace(\"BUSpaRse\", quietly = TRUE)) BiocManager::install(\"BUSpaRse\", version = bioconductor_version, update = FALSE)\n",
    "if (!requireNamespace(\"DropletUtils\", quietly = TRUE)) BiocManager::install(\"DropletUtils\", version = bioconductor_version, update = FALSE)\n",
    "if (!requireNamespace(\"biomaRt\", quietly = TRUE)) BiocManager::install(\"biomaRt\", version = bioconductor_version, update = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0ea1477",
   "metadata": {},
   "outputs": [],
   "source": [
    "if (seu1_data_path == \"\") {\n",
    "    seu1_data_path <- glue::glue(\"{data_path_root}/seuratv5/seu.rds\")\n",
    "}\n",
    "\n",
    "if (seu2_data_path == \"\") {\n",
    "    seu2_data_path <- glue::glue(\"{data_path_root}/seuratv4/seu.rds\")\n",
    "}\n",
    "\n",
    "if (seu1_markers_data_path == \"\") {\n",
    "    seu1_markers_data_path <- glue::glue(\"{data_path_root}/seuratv5/markers.rds\")\n",
    "}\n",
    "\n",
    "if (seu2_markers_data_path == \"\") {\n",
    "    seu2_markers_data_path <- glue::glue(\"{data_path_root}/seuratv4/markers.rds\")\n",
    "}\n",
    "\n",
    "if (output_base_path == \"\") {\n",
    "    output_base_path <- glue::glue(\"{project_base_path}/output/{data_name}/seuratv{seurat1_version}vs_seuratv{seurat2_version}\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7124a57",
   "metadata": {},
   "source": [
    "R Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75ae83db",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_group_names <- list(Seurat1 = seu1_name, Seurat2 = seu2_name)\n",
    "\n",
    "group1_color <- \"#009E73\"\n",
    "group2_color <- \"#CC79A7\"\n",
    "\n",
    "conda_env <- \"analysis_env\"\n",
    "Sys.setenv(RETICULATE_PYTHON = paste(\"/home/rstudio/.conda/envs\", conda_env, \"bin/python3.9\", sep = \"/\"))\n",
    "library(reticulate)\n",
    "use_condaenv(conda_env)\n",
    "library(Seurat)\n",
    "library(Matrix)\n",
    "library(tidyverse)\n",
    "library(patchwork)\n",
    "library(eulerr)\n",
    "library(scattermore)\n",
    "library(DropletUtils)\n",
    "library(glue)\n",
    "library(bluster)\n",
    "library(ggforce)\n",
    "library(ggplotify)\n",
    "library(grid)\n",
    "library(gtable)\n",
    "library(ggalluvial)\n",
    "theme_set(theme_bw())\n",
    "\n",
    "source(glue(\"{project_base_path}/scripts/data_analysis_helper.R\"))\n",
    "source(glue(\"{project_base_path}/scripts/plotting_and_stats.R\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d2e517d",
   "metadata": {},
   "source": [
    "Download data if necessary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94de6647",
   "metadata": {},
   "outputs": [],
   "source": [
    "py_run_string('import sys\n",
    "sys.path.append(f\"{r.project_base_path}/scripts\")\n",
    "from download_data import *\n",
    "\n",
    "if r.download_data:\n",
    "    download_and_extract(r.doi, r.data_name, r.data_path_root)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d201579",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_hvg_flavor <- \"vst\"\n",
    "seu_mean_cutoff <- c(0.1, 8)\n",
    "seu_dispersion_cutoff <- c(1, Inf)\n",
    "seu_vars_to_regress <- NULL\n",
    "seurat_scale_max <- 10\n",
    "seu_n_neighbors <- 20\n",
    "seurat_clustering_algorithm <- \"louvain\"\n",
    "seu_resolution <- 0.8\n",
    "seu_umap_min_dist <- 0.5\n",
    "seurat_umap_seed <- 0\n",
    "seu_umap_method <- \"uwot\"\n",
    "seu_umap_metric <- \"correlation\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00bf406f",
   "metadata": {},
   "source": [
    "Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f048ba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "seu1 <- readRDS(seu1_data_path)\n",
    "markers_seu1 <- readRDS(seu1_markers_data_path)\n",
    "\n",
    "seu2 <- readRDS(seu2_data_path)\n",
    "markers_seu2 <- readRDS(seu2_markers_data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f673a337",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_data_file_paths <- list(\n",
    "    markers_seu1 = glue::glue(\"{output_base_path}/data_files/markers_{seurat1_version}.rds\"),\n",
    "    markers_seu2 = glue::glue(\"{output_base_path}/data_files/markers_{seurat2_version}.rds\"),\n",
    "    markers2 = glue::glue(\"{output_base_path}/data_files/markers2.rds\"),\n",
    "    seu1_object = glue::glue(\"{output_base_path}/data_files/seu1.rds\"),\n",
    "    seu2_object = glue::glue(\"{output_base_path}/data_files/seu2.rds\")\n",
    ")\n",
    "\n",
    "# FALSE to have no save\n",
    "file_paths <- list(\n",
    "    filter_arguments = glue::glue(\"{output_base_path}/stats/filter_stats.txt\"),\n",
    "    euler_stats_before_QC_file = FALSE, # glue::glue(\"{output_base_path}/stats/euler_stats_beforeQC.txt\"),\n",
    "    euler_stats_after_QC_file = glue::glue(\"{output_base_path}/stats/euler_stats_afterQC.txt\"),\n",
    "    pca_knn_clustering_umap_file = glue::glue(\"{output_base_path}/stats/pca_knn_clustering_umap_stats.txt\"),\n",
    "    de_stats_file = glue::glue(\"{output_base_path}/stats/de_stats.txt\"),\n",
    "    upset_cells = glue::glue(\"{output_base_path}/plots/upset_cells.tiff\"),\n",
    "    upset_genes = glue::glue(\"{output_base_path}/plots/upset_genes.tiff\"),\n",
    "    upset_hvgs = glue::glue(\"{output_base_path}/plots/upset_hvgs.tiff\"),\n",
    "    upset_markers_genes_only = glue::glue(\"{output_base_path}/plots/upset_marker_genes_only.tiff\"),\n",
    "    upset_markers = glue::glue(\"{output_base_path}/plots/upset_markers.tiff\"),\n",
    "    pca_12_overlay_filepath = glue::glue(\"{output_base_path}/plots/pca_scatterplot_12.tiff\"),\n",
    "    combined_pc_variance_loadings_plot = glue::glue(\"{output_base_path}/plots/combined_pc_variance_loadings_plot.tiff\"),\n",
    "    jaccard_degree_scatterplot = glue::glue(\"{output_base_path}/plots/jaccard_degree_scatterplot.tiff\"),\n",
    "    alluvial = glue::glue(\"{output_base_path}/plots/cluster_alluvial.tiff\"),\n",
    "    alluvial_legend = glue::glue(\"{output_base_path}/plots/cluster_alluvial_legend.tiff\"),\n",
    "    umap_seu1 = glue::glue(\"{output_base_path}/plots/umap_seu_{seu1_name}.tiff\"),\n",
    "    umap_seu2 = glue::glue(\"{output_base_path}/plots/umap_seu_{seu2_name}.tiff\"),\n",
    "    umap_jaccard_degree_scatterplot = glue::glue(\"{output_base_path}/plots/umap_jaccard_degree_scatterplot.tiff\"),\n",
    "    umap_jaccard_knn_density = glue::glue(\"{output_base_path}/plots/umap_jaccard_knn_density.tiff\"),\n",
    "    umap_jaccard_knn_density_seu1_facet = glue::glue(\"{output_base_path}/plots/umap_jaccard_knn_density_seu1_facet.tiff\"),\n",
    "    umap_jaccard_knn_density_seu2_facet = glue::glue(\"{output_base_path}/plots/umap_jaccard_knn_density_seu2_facet.tiff\"),\n",
    "    umap_alluvial = glue::glue(\"{output_base_path}/plots/umap_alluvial.tiff\"),\n",
    "    umap_umap_leiden_seu1 = glue::glue(\"{output_base_path}/plots/umap_umap_leiden_seu1.tiff\"),\n",
    "    umap_umap_leiden_seu2 = glue::glue(\"{output_base_path}/plots/umap_umap_leiden_seu2.tiff\"),\n",
    "    logFC_scatterplot_file_path = glue::glue(\"{output_base_path}/plots/logFC_scatterplot.tiff\"),\n",
    "    wilcoxon_scatterplot_file_path = glue::glue(\"{output_base_path}/plots/wilcoxon_scatterplot.tiff\"),\n",
    "    logFC_scatterplot_file_path_with_legend = glue::glue(\"{output_base_path}/plots/logFC_scatterplot_with_legend.tiff\"),\n",
    "    wilcoxon_scatterplot_file_path_with_legend = glue::glue(\"{output_base_path}/plots/wilcoxon_scatterplot_with_legend.tiff\")\n",
    ")\n",
    "\n",
    "for (path in output_data_file_paths) {\n",
    "    dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)\n",
    "}\n",
    "\n",
    "\n",
    "if (save_data) {\n",
    "    for (path in output_data_file_paths) {\n",
    "        dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)\n",
    "    }\n",
    "    \n",
    "    for (path in file_paths) {\n",
    "        if (is.character(path)) {\n",
    "            # Extract the directory part of the path\n",
    "            specific_output_path <- dirname(path)\n",
    "\n",
    "            # Create the directory if it does not exist\n",
    "            if (!dir.exists(specific_output_path)) {\n",
    "                dir.create(specific_output_path, recursive = TRUE, showWarnings = FALSE)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    for (file in c(file_paths$euler_stats_after_QC_file, file_paths$pca_knn_clustering_umap_file, file_paths$de_stats_file)) {\n",
    "        if (is.character(file)) {\n",
    "            sink(file = file, append = FALSE)\n",
    "            sink()\n",
    "        }\n",
    "    }\n",
    "} else {\n",
    "    for (i in seq_along(file_paths)) {\n",
    "        file_paths[[i]] <- FALSE\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c587d4a",
   "metadata": {},
   "source": [
    "Overlaps, PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c87e71f",
   "metadata": {},
   "outputs": [],
   "source": [
    "upset_cell <- make_upset_seurat(group1 = seu1, group2 = seu2, comparison = \"Cell\", group_names = seurat_group_names, save = file_paths$upset_cells)\n",
    "upset_gene <- make_upset_seurat(group1 = seu1, group2 = seu2, comparison = \"Gene\", group_names = seurat_group_names, save = file_paths$upset_genes)\n",
    "upset_cell\n",
    "upset_gene\n",
    "\n",
    "upset_hvg <- make_upset_seurat(seu1, seu2, comparison = \"HVG\", group_names = seurat_group_names, save = file_paths$upset_hvgs)\n",
    "upset_hvg\n",
    "\n",
    "tot_variance1 <- Misc(Reductions(seu1, \"pca\"))[[\"total.variance\"]]\n",
    "var_explained1 <- Stdev(seu1, reduction = \"pca\")^2 / tot_variance1\n",
    "\n",
    "tot_variance2 <- Misc(Reductions(seu2, \"pca\"))[[\"total.variance\"]]\n",
    "var_explained2 <- Stdev(seu2, reduction = \"pca\")^2 / tot_variance2\n",
    "\n",
    "eigs_df <- tibble(\n",
    "    Seurat1 = var_explained1,\n",
    "    Seurat2 = var_explained2,\n",
    "    PC = 1:50\n",
    ")\n",
    "\n",
    "combined_pc_variance <- plot_var_explained(eigs_df, npcs = 50, group_names = unlist(seurat_group_names), save = FALSE)\n",
    "\n",
    "pca_embeddings1 <- Embeddings(seu1, reduction = \"pca\")\n",
    "pca_embeddings2 <- Embeddings(seu2, reduction = \"pca\")\n",
    "\n",
    "pca12_plot <- plot_pca_compare(pca_embeddings1, pca_embeddings2, group1_name = \"Seurat1\", group2_name = \"Seurat2\", group_labels = unlist(seurat_group_names), save = file_paths$pca_12_overlay_filepath)\n",
    "pca12_plot\n",
    "\n",
    "pca_loadings_seu1 <- Loadings(seu1, reduction = \"pca\")\n",
    "pca_loadings_seu2 <- Loadings(seu2, reduction = \"pca\")\n",
    "\n",
    "df_loadings <- make_pc_diffs_df(list(\n",
    "    Seurat1 = pca_loadings_seu1,\n",
    "    Seurat2 = pca_loadings_seu2\n",
    "), npcs = 50)\n",
    "\n",
    "mean_loadings_diff <- mean(df_loadings$differences[1:3])\n",
    "\n",
    "if (file_paths$pca_knn_clustering_umap_file != FALSE) {\n",
    "    sink(file_paths$pca_knn_clustering_umap_file, split = TRUE, append = TRUE)\n",
    "}\n",
    "\n",
    "print(glue(\"Mean loading difference of PC1-3: {mean_loadings_diff}\"))\n",
    "\n",
    "if (file_paths$pca_knn_clustering_umap_file != FALSE) {\n",
    "    sink()\n",
    "}\n",
    "\n",
    "mylist <- list(\n",
    "    Seurat1 = pca_loadings_seu1,\n",
    "    Seurat2 = pca_loadings_seu2\n",
    ")\n",
    "\n",
    "loading_diffs <- plot_loading_diffs(df_loadings, save = FALSE)\n",
    "\n",
    "\n",
    "combined_plot <- make_combined_pc_variance_loadings_plot(combined_pc_variance, loading_diffs, save = file_paths$combined_pc_variance_loadings_plot)\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a261f89",
   "metadata": {},
   "source": [
    "KNN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78348d14",
   "metadata": {},
   "outputs": [],
   "source": [
    "seu1_inds <- colnames(seu1)\n",
    "seu2_inds <- colnames(seu2)\n",
    "\n",
    "snn_graph_seu1 <- seu1@graphs$RNA_snn\n",
    "snn_graph_seu2 <- seu2@graphs$RNA_snn\n",
    "\n",
    "seu_snn_b1 <- snn_graph_seu1 > 0\n",
    "seu_snn_b2 <- snn_graph_seu2 > 0\n",
    "\n",
    "if (!identical(seu1_inds, seu2_inds)) {\n",
    "    seu_snn_b1 <- seu_snn_b1[overlapping_inds, overlapping_inds]\n",
    "    seu_snn_b2 <- seu_snn_b2[overlapping_inds, overlapping_inds]\n",
    "}\n",
    "\n",
    "seu1_list <- mat2list(seu_snn_b1)\n",
    "seu2_list <- mat2list(seu_snn_b2)\n",
    "\n",
    "jaccards <- find_jaccards(list(Seurat1 = seu1_list, Seurat2 = seu2_list))\n",
    "\n",
    "median_jaccard <- median(jaccards$Jaccard)\n",
    "\n",
    "jaccard_plot <- make_jaccard_plot(jaccards, median_jaccard, save = FALSE)\n",
    "\n",
    "jaccard_plot\n",
    "\n",
    "nei_sizes <- tibble(\n",
    "    Seurat1 = lengths(seu1_list),\n",
    "    Seurat2 = lengths(seu2_list)\n",
    ")\n",
    "\n",
    "nei_pairs <- make_pairwise_df(nei_sizes)\n",
    "\n",
    "knn_scatterplot <- make_knn_scatterplot(nei_pairs, save = FALSE)\n",
    "\n",
    "knn_scatterplot\n",
    "\n",
    "jaccards$degree_ratio <- nei_pairs$value1 / nei_pairs$value2\n",
    "jaccards$logged_degree_ratio <- log(jaccards$degree_ratio, base = 2)\n",
    "\n",
    "jaccards$jaccard_logged <- log(jaccards$Jaccard, base = 2)\n",
    "\n",
    "median_logged_degree_ratio <- median(jaccards$logged_degree_ratio)\n",
    "\n",
    "if (file_paths$pca_knn_clustering_umap_file != FALSE) {\n",
    "    sink(file_paths$pca_knn_clustering_umap_file, append = TRUE, split = TRUE)\n",
    "}\n",
    "\n",
    "print(glue(\"Median jaccard of KNN: {median_jaccard}\"))\n",
    "print(glue(\"Median log degree ratio of KNN: {median_logged_degree_ratio}\"))\n",
    "\n",
    "if (file_paths$pca_knn_clustering_umap_file != FALSE) {\n",
    "    sink()\n",
    "}\n",
    "\n",
    "\n",
    "jaccard_degree_scatterplot <- make_snn_jaccard_degree_scatterplot(jaccards, save = file_paths$jaccard_degree_scatterplot)\n",
    "jaccard_degree_scatterplot\n",
    "\n",
    "seu1_clusters <- Idents(seu1)\n",
    "seu2_clusters <- Idents(seu2)\n",
    "\n",
    "seu1_clusters_vector <- as.vector(seu1_clusters)\n",
    "seu2_clusters_vector <- as.vector(seu2_clusters)\n",
    "ari_value <- mclust::adjustedRandIndex(seu1_clusters_vector, seu2_clusters_vector)\n",
    "\n",
    "\n",
    "if (file_paths$pca_knn_clustering_umap_file != FALSE) {\n",
    "    sink(file_paths$pca_knn_clustering_umap_file, append = TRUE, split = TRUE)\n",
    "}\n",
    "\n",
    "print(glue(\"Adjusted Rand index between clusters: {ari_value}\"))\n",
    "\n",
    "if (file_paths$pca_knn_clustering_umap_file != FALSE) {\n",
    "    sink()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a94d450",
   "metadata": {},
   "source": [
    "Clustering, UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4306aab",
   "metadata": {},
   "outputs": [],
   "source": [
    "seu1_clusters <- Idents(seu1)\n",
    "seu2_clusters <- Idents(seu2)\n",
    "\n",
    "df <- tibble(\n",
    "    Seurat1 = seu1_clusters,\n",
    "    Seurat2 = seu2_clusters\n",
    ")\n",
    "\n",
    "df <- setNames(df, unlist(seurat_group_names))\n",
    "\n",
    "clus_df_gather <- get_alluvial_df(df)\n",
    "\n",
    "clus_df_gather2 <- sort_clusters_by_agreement(clus_df_gather, stable_column = seu1_name, reordered_column = seu2_name)\n",
    "\n",
    "alluvial_plot <- plot_alluvial(clus_df_gather, color_boxes = TRUE, color_bands = FALSE, group1_name = seu1_name, group2_name = seu2_name, save = file_paths$alluvial)\n",
    "alluvial_plot_legend <- plot_alluvial(clus_df_gather, color_boxes = TRUE, color_bands = TRUE, alluvial_alpha = 0.5, group1_name = seu1_name, group2_name = seu2_name, save = file_paths$alluvial_legend)\n",
    "alluvial_plot\n",
    "alluvial_plot_legend\n",
    "\n",
    "seu1_umap_info <- RunUMAP(seu1, dims = 1:50, min.dist = seu_umap_min_dist, umap.method = seu_umap_method, seed.use = 42, metric = seu_umap_metric)\n",
    "seu2_umap_info <- RunUMAP(seu2, dims = 1:50, min.dist = seu_umap_min_dist, umap.method = seu_umap_method, seed.use = 42, metric = seu_umap_metric)\n",
    "\n",
    "colors_group2 <- find_group2_colors(clus_df_gather, seu1_name, seu2_name)\n",
    "\n",
    "umap_plots <- plot_umap(group1_umap_info = seu1_umap_info$umap@cell.embeddings, group1_clusters = seu1$seurat_clusters, group2_umap_info = seu2_umap_info$umap@cell.embeddings, group2_clusters = seu2$seurat_clusters, colors_group2 = colors_group2, group1 = seu1_name, group2 = seu2_name, save = c(file_paths$umap_seu1, file_paths$umap_seu2))\n",
    "seu1_umap <- umap_plots[[1]]\n",
    "seu2_umap <- umap_plots[[2]]\n",
    "\n",
    "seu1_umap\n",
    "seu2_umap\n",
    "\n",
    "seu1_umap_data <- seu1_umap_info$umap@cell.embeddings\n",
    "seu2_umap_data <- seu2_umap_info$umap@cell.embeddings\n",
    "\n",
    "seu1_umap_knn <- dbscan::kNN(seu1_umap_data, k = 50)\n",
    "seu2_umap_knn <- dbscan::kNN(seu2_umap_data, k = 50)\n",
    "\n",
    "jaccards_all_cells <- calculate_knn_jaccards(seu1_umap_knn$id, seu2_umap_knn$id)\n",
    "\n",
    "median_jaccard_umap_knn <- median(jaccards_all_cells)\n",
    "\n",
    "if (file_paths$pca_knn_clustering_umap_file != FALSE) {\n",
    "    sink(file_paths$pca_knn_clustering_umap_file, append = TRUE, split = TRUE)\n",
    "}\n",
    "\n",
    "print(glue(\"Median jaccard of UMAP KNN: {median_jaccard_umap_knn}\"))\n",
    "\n",
    "if (file_paths$pca_knn_clustering_umap_file != FALSE) {\n",
    "    sink()\n",
    "}\n",
    "\n",
    "overlapping_inds <- intersect(seu1_inds, seu2_inds)\n",
    "\n",
    "seu1_cluster_data_filtered <- Idents(seu1)[overlapping_inds]\n",
    "seu2_cluster_data_filtered <- Idents(seu2)[overlapping_inds]\n",
    "\n",
    "jaccards_df <- data.frame(Cells = overlapping_inds, JaccardIndex = jaccards_all_cells, seu1_clusters = seu1_cluster_data_filtered, seu2_clusters = seu2_cluster_data_filtered)\n",
    "\n",
    "umap_jaccard_plot <- make_umap_jaccard_plot(jaccards_df, save = file_paths$umap_jaccard_knn_density)\n",
    "\n",
    "umap_jaccard_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5fcb211",
   "metadata": {},
   "source": [
    "DE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5526ee84",
   "metadata": {},
   "outputs": [],
   "source": [
    "seu1_filtered_markers <- markers_seu1 %>% filter(p_val_adj < 0.05)\n",
    "seu2_filtered_markers <- markers_seu2 %>% filter(p_val_adj < 0.05)\n",
    "\n",
    "# vectorized_seu_unfiltered_markers <- unique(markers$gene)\n",
    "vectorized_seu1_filtered_markers <- unique(seu1_filtered_markers$gene)\n",
    "vectorized_seu2_filtered_markers <- unique(seu2_filtered_markers$gene)\n",
    "\n",
    "\n",
    "markers_euler_genes_only <- make_euler_seurat(vectorized_seu1_filtered_markers, vectorized_seu2_filtered_markers, comparison = \"Marker Gene\", group_names = seurat_group_names, save_plot = FALSE, save_stats = FALSE)\n",
    "markers_euler_genes_only\n",
    "\n",
    "upset_marker_gene_only <- make_upset_seurat(vectorized_seu1_filtered_markers, vectorized_seu2_filtered_markers, comparison = \"Marker Gene\", group_names = seurat_group_names, save = file_paths$upset_markers_genes_only)\n",
    "\n",
    "# From this point on, make sure to use markers dataframes with aligned cluster info\n",
    "if (!(identical(seu1_clusters, seu2_clusters))) {\n",
    "    stop(\"The groups have unequal cell sets, so not running further DE analysis, which requires clusters to be in agreement.\")\n",
    "}\n",
    "\n",
    "seu1_markers_df <- markers_seu1 %>% select(gene = gene, cluster = cluster)\n",
    "seu2_markers_df <- markers_seu2 %>% select(gene = gene, cluster = cluster)\n",
    "\n",
    "vectorized_seu1_markers <- paste(seu1_markers_df$gene, seu1_markers_df$cluster, sep = \"-\")\n",
    "vectorized_seu2_markers <- paste(seu2_markers_df$gene, seu2_markers_df$cluster, sep = \"-\")\n",
    "\n",
    "markers_euler <- make_euler_seurat(vectorized_seu1_markers, vectorized_seu2_markers, comparison = \"Marker\", group_names = seurat_group_names, save_plot = FALSE, save_stats = FALSE)\n",
    "markers_euler\n",
    "\n",
    "upset_markers_all <- make_upset_seurat(vectorized_seu1_markers, vectorized_seu2_markers, comparison = \"Marker\", group_names = seurat_group_names, save = file_paths$upset_markers)\n",
    "\n",
    "markers2 <- markers_seu1 |>\n",
    "    inner_join(markers_seu2, by = c(\"cluster\", \"gene\"), suffix = c(glue(\".{seu1_name}\"), glue(\".{seu2_name}\")))\n",
    "\n",
    "markers2 <- markers2 |>\n",
    "    mutate(cluster = factor(cluster, levels = as.character(seq_len(length(unique(cluster))) - 1)))\n",
    "\n",
    "markers2 <- markers2 |>\n",
    "    group_by(cluster) |>\n",
    "    mutate(rank_r = seq_along(gene))\n",
    "\n",
    "\n",
    "markers2 <- calculate_de_stats(markers2, group1_name = seu1_name, group2_name = seu2_name, save = FALSE)\n",
    "\n",
    "source(glue(\"{project_base_path}/scripts/plotting_and_stats.R\"))\n",
    "markers2[[glue(\"p_val_adj.{seu1_name}\")]][markers2[[glue(\"p_val_adj.{seu1_name}\")]] == 0] <- .Machine$double.xmin\n",
    "markers2[[glue(\"p_val_adj.{seu2_name}\")]][markers2[[glue(\"p_val_adj.{seu2_name}\")]] == 0] <- .Machine$double.xmin\n",
    "\n",
    "logFC_scatterplot <- plot_scatterplot_de_logfc(markers2, group1_name = seu1_name, group2_name = seu2_name, save = file_paths$logFC_scatterplot_file_path, outliers_excluded = FALSE)\n",
    "pvaladj_scatterplot <- plot_scatterplot_de_wilcoxon(markers2, group1_name = seu1_name, group2_name = seu2_name, save = file_paths$wilcoxon_scatterplot_file_path, outliers_excluded = FALSE)\n",
    "\n",
    "logFC_scatterplot_with_legend <- plot_scatterplot_de_logfc(markers2, group1_name = seu1_name, group2_name = seu2_name, save = file_paths$logFC_scatterplot_file_path_with_legend, outliers_excluded = FALSE, show_legend = TRUE)\n",
    "\n",
    "logFC_scatterplot\n",
    "logFC_scatterplot_with_legend\n",
    "\n",
    "pvaladj_scatterplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2be8f8c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(markers2, file = output_data_file_paths$markers2)"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
